// @bitgo/sdk-api/dist/src/encrypt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.decrypt = exports.encrypt = exports.bytesToWord = void 0;
const sjcl = require("@bitgo/sjcl");
const crypto_1 = require("crypto");
/**
 * convert a 4 element Uint8Array to a 4 byte Number
 *
 * @param bytes
 * @return 4 byte number
 */
function bytesToWord(bytes) {
    if (!(bytes instanceof Uint8Array) || bytes.length !== 4) {
        throw new Error('bytes must be a Uint8Array with length 4');
    }
    return bytes.reduce((num, byte) => num * 0x100 + byte, 0);
}
exports.bytesToWord = bytesToWord;
function encrypt(password, plaintext, options) {
    const salt = (options === null || options === void 0 ? void 0 : options.salt) || (0, crypto_1.randomBytes)(8);
    if (salt.length !== 8) {
        throw new Error(`salt must be 8 bytes`);
    }
    const iv = (options === null || options === void 0 ? void 0 : options.iv) || (0, crypto_1.randomBytes)(16);
    if (iv.length !== 16) {
        throw new Error(`iv must be 16 bytes`);
    }
    const encryptOptions = {
        iter: 10000,
        ks: 256,
        salt: [bytesToWord(salt.slice(0, 4)), bytesToWord(salt.slice(4))],
        iv: [
            bytesToWord(iv.slice(0, 4)),
            bytesToWord(iv.slice(4, 8)),
            bytesToWord(iv.slice(8, 12)),
            bytesToWord(iv.slice(12, 16)),
        ],
    };
    if (options === null || options === void 0 ? void 0 : options.adata) {
        encryptOptions.adata = options.adata;
    }
    return sjcl.encrypt(password, plaintext, encryptOptions);
}
exports.encrypt = encrypt;
function decrypt(password, ciphertext) {
    return sjcl.decrypt(password, ciphertext);
}
exports.decrypt = decrypt;
