// @bitgo/statics/dist/src/base.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseCoin = exports.BaseUnit = exports.KeyCurve = exports.UnderlyingAsset = exports.CoinFeature = exports.CoinFamily = exports.CoinKind = void 0;
const errors_1 = require("./errors");
var CoinKind;
(function (CoinKind) {
    CoinKind["CRYPTO"] = "crypto";
    CoinKind["FIAT"] = "fiat";
})(CoinKind = exports.CoinKind || (exports.CoinKind = {}));
/**
 * The coin family links related variants of a single coin together.
 *
 * Typically, each coin will have a testnet and mainnet variant,
 * and these will both have the same coin family.
 *
 * For example, the coins `btc` and `tbtc` both belong to the same family, `btc`.
 */
var CoinFamily;
(function (CoinFamily) {
    CoinFamily["ADA"] = "ada";
    CoinFamily["ALGO"] = "algo";
    CoinFamily["ARBETH"] = "arbeth";
    CoinFamily["ATOM"] = "atom";
    CoinFamily["AVAXC"] = "avaxc";
    CoinFamily["AVAXP"] = "avaxp";
    CoinFamily["BASEETH"] = "baseeth";
    CoinFamily["BCH"] = "bch";
    CoinFamily["BCHA"] = "bcha";
    CoinFamily["BERA"] = "bera";
    CoinFamily["BLD"] = "bld";
    CoinFamily["BSC"] = "bsc";
    CoinFamily["BSV"] = "bsv";
    CoinFamily["BTC"] = "btc";
    CoinFamily["BTG"] = "btg";
    CoinFamily["CELO"] = "celo";
    CoinFamily["COREUM"] = "coreum";
    CoinFamily["CSPR"] = "cspr";
    CoinFamily["DASH"] = "dash";
    CoinFamily["DOGE"] = "doge";
    CoinFamily["DOT"] = "dot";
    CoinFamily["ETH"] = "eth";
    CoinFamily["ETH2"] = "eth2";
    CoinFamily["ETHW"] = "ethw";
    CoinFamily["ETC"] = "etc";
    CoinFamily["EOS"] = "eos";
    CoinFamily["FIAT"] = "fiat";
    CoinFamily["HASH"] = "hash";
    CoinFamily["HBAR"] = "hbar";
    CoinFamily["INJECTIVE"] = "injective";
    CoinFamily["ISLM"] = "islm";
    CoinFamily["KAVA"] = "kava";
    CoinFamily["LNBTC"] = "lnbtc";
    CoinFamily["LTC"] = "ltc";
    CoinFamily["POLYGON"] = "polygon";
    CoinFamily["NEAR"] = "near";
    CoinFamily["OFC"] = "ofc";
    CoinFamily["OPETH"] = "opeth";
    CoinFamily["OSMO"] = "osmo";
    CoinFamily["RBTC"] = "rbtc";
    CoinFamily["SEI"] = "sei";
    CoinFamily["SOL"] = "sol";
    CoinFamily["SUI"] = "sui";
    CoinFamily["STX"] = "stx";
    CoinFamily["SUSD"] = "susd";
    CoinFamily["TIA"] = "tia";
    CoinFamily["TON"] = "ton";
    CoinFamily["TRX"] = "trx";
    CoinFamily["XLM"] = "xlm";
    CoinFamily["XRP"] = "xrp";
    CoinFamily["XTZ"] = "xtz";
    CoinFamily["ZEC"] = "zec";
    CoinFamily["ZETA"] = "zeta";
    CoinFamily["ZKETH"] = "zketh";
})(CoinFamily = exports.CoinFamily || (exports.CoinFamily = {}));
/**
 * Coin features are yes or no questions about what a coin requires or is capable of.
 *
 * This allows coin-agnostic handling of coin-specific features. This is designed
 * to replace checking the coin name against a whitelist of supported coins
 * before executing some coin-specific logic, and instead allows one to check if a
 * coin supports the coin-specific feature that the logic implements.
 */
var CoinFeature;
(function (CoinFeature) {
    /*
     * This coin supports creating wallets on different networks with the same keys. Only works for TSS account-base coins
     */
    CoinFeature["EVM_WALLET"] = "evm-wallet";
    /*
     * This coin supports creating an EVM transaction using Metamask Institutional (MMI).
     */
    CoinFeature["METAMASK_INSTITUTIONAL"] = "metamask-institutional";
    /*
     * The valueless transfer feature indicates that it is valid to send a transaction which moves zero units of the coin.
     *
     * An example is Ethereum, which uses zero value transactions to trigger contract calls.
     */
    CoinFeature["VALUELESS_TRANSFER"] = "valueless-transfer";
    /*
     * Transaction data means there can be arbitrary data encoded in a transaction.
     *
     * Ethereum contract call data is an example.
     */
    CoinFeature["TRANSACTION_DATA"] = "transaction-data";
    /*
     * Some coins have a higher precision range than IEEE 754 doubles, which are used to represent numbers in javascript.
     *
     * For these coins, we must use an arbitrary precision arithmetic library, and this feature indicates this requirement.
     */
    CoinFeature["REQUIRES_BIG_NUMBER"] = "requires-big-number";
    /*
     * RMG requires all wallets to have a backup key held by a BitGo approved Key Recovery Service (KRS)
     */
    CoinFeature["REQUIRES_KRS_BACKUP_KEY"] = "requires-krs-backup-key";
    /*
     * For customers which are not on a postpaid contract, we add an extra output to transactions which pays BitGo a fee.
     *
     * This fee is known as the "pay-as-you-go fee", or just "paygo" for short.
     *
     * Some coins are unable to create transactions with more than one output, so paygo outputs are not possible for these coins.
     */
    CoinFeature["PAYGO"] = "paygo";
    /*
     * Does this coin align with the unspent model?
     *
     * These are typically Bitcoin and forks of it, such as Litecoin and Bitcoin Cash.
     */
    CoinFeature["UNSPENT_MODEL"] = "unspent-model";
    /*
     * Does this coin align with the Lightning Network model?
     *
     * These are typically Lightning Network on unspent model coins, such as BTC and LBTC.
     */
    CoinFeature["LIGHTNING_MODEL"] = "lightning-model";
    /*
     * Does this coin align with the account model?
     *
     * Examples of this coin type are Ethereum, XRP, and Stellar
     */
    CoinFeature["ACCOUNT_MODEL"] = "account-model";
    /*
     * Does this coin support child-pays-for-parent transactions?
     *
     * These are special types of transactions which can accelerate the confirmation time
     * of another transaction which is stuck in the mempool due to low fees.
     *
     * This is only possible for coins which follow the unspent model (UTXO coins).
     */
    CoinFeature["CHILD_PAYS_FOR_PARENT"] = "cpfp";
    /*
     * Does this coin support tokens? These are distinct assets from the underlying coin, but run on the same network.
     *
     * For example, Ethereum's ERC 20 token standard means that it supports tokens, so it shall have this feature.
     */
    CoinFeature["SUPPORTS_TOKENS"] = "supports-tokens";
    /*
     * Are fees for transactions of this coin paid for by the Enterprise (eg, Enterprise gas tank)?
     */
    CoinFeature["ENTERPRISE_PAYS_FEES"] = "enterprise-pays-fees";
    /*
     * This coin requires that accounts keep a minimum balance as reserve
     */
    CoinFeature["REQUIRES_RESERVE"] = "requires-reserve";
    /**
     * @deprecated This property is no longer valid. Please select the following custody option based on the BitGo org:
     * * CUSTODY_BITGO_TRUST
     * * CUSTODY_BITGO_NEW_YORK
     * * CUSTODY_BITGO_GERMANY
     * * CUSTODY_BITGO_SWITZERLAND
     */
    CoinFeature["CUSTODY"] = "custody";
    /*
    This coin uses TSS for key creation and signing
     */
    CoinFeature["TSS"] = "tss";
    /*
     * This coin supports staking
     */
    CoinFeature["STAKING"] = "staking";
    /**
     * This coin is deprecated
     */
    CoinFeature["DEPRECATED"] = "deprecated";
    /**
     * This coin is a dummy object meant to be a placeholder for an unsupported token
     */
    CoinFeature["GENERIC_TOKEN"] = "genericToken";
    /*
     * This coin supports custody in BitGo Trust SD entities
     */
    CoinFeature["CUSTODY_BITGO_TRUST"] = "custody-bitgo-trust";
    /*
     * This coin supports custody in BitGo New York entities
     */
    CoinFeature["CUSTODY_BITGO_NEW_YORK"] = "custody-bitgo-new-york";
    /*
     * This coin supports custody in BitGo Germany entities
     */
    CoinFeature["CUSTODY_BITGO_GERMANY"] = "custody-bitgo-germany";
    /*
     * This coin supports custody in BitGo Switzerland entities
     */
    CoinFeature["CUSTODY_BITGO_SWITZERLAND"] = "custody-bitgo-switzerland";
    /*
     * This coin supports custody in BitGo Switzerland entities
     */
    CoinFeature["CUSTODY_BITGO_FRANKFURT"] = "custody-bitgo-frankfurt";
    /*
     * This coin supports custody in BitGo Singapore entities
     */
    CoinFeature["CUSTODY_BITGO_SINGAPORE"] = "custody-bitgo-singapore";
    /*
     * This coin supports custody in BitGo Sister Trust 1 entities
     */
    CoinFeature["CUSTODY_BITGO_SISTER_TRUST_ONE"] = "custody-bitgo-sister-trust-one";
    /*
     * This coin has transactions that expire after a certain amount of time.
     */
    CoinFeature["EXPIRING_TRANSACTIONS"] = "expiring-transactions";
    /**
     * This coin supports cold wallets that use a multisig signing protocol
     */
    CoinFeature["MULTISIG_COLD"] = "multisig-cold";
    /**
     * This coin supports cold wallets that use a TSS signing protocol
     */
    CoinFeature["TSS_COLD"] = "tss-cold";
    /**
     * This coin uses sha256 hash function for ECDSA TSS signatures
     */
    CoinFeature["SHA256_WITH_ECDSA_TSS"] = "sha256-with-ecdsa-tss";
    /**
     * This coin is cosmos like coin
     */
    CoinFeature["COSMOS_LIKE_COINS"] = "cosmos_like_coins";
    /**
     * This coin supports the ability to rebuild transactions on custody signing
     */
    CoinFeature["REBUILD_ON_CUSTODY_SIGNING"] = "rebuild-on-custody-signing";
    /**
     * This coin supports higher limit for tx request rebuild, which is 10 by default
     */
    CoinFeature["INCREASED_TX_REQUEST_REBUILD_LIMIT"] = "increased-tx-request-rebuild-limit";
    /**
     * This coin supports bulk transaction creation
     */
    CoinFeature["BULK_TRANSACTION"] = "bulk-transaction";
    /**
     * This coin supports distributed custody wallets
     */
    CoinFeature["DISTRIBUTED_CUSTODY"] = "distributed-custody";
    /**
     * This coin supports bulk staking transaction creation
     */
    CoinFeature["BULK_STAKING_TRANSACTION"] = "bulk-staking-transaction";
    /**
     * This coin uses non-packed encoding for transaction data
     */
    CoinFeature["USES_NON_PACKED_ENCODING_FOR_TXDATA"] = "uses-non-packed-encoding-for-txdata";
    /**
     * This coins supports MPCv2 for key creation and signing
     */
    CoinFeature["MPCV2"] = "mpcv2";
    /**
     * This coin supports acceleration or nonce filling txn for stuck transactions for tss wallet
     */
    CoinFeature["STUCK_TRANSACTION_MANAGEMENT_TSS"] = "stuck-transaction-management-tss";
    /**
     * This coin supports acceleration or nonce filling txn for stuck transactions for onchain wallet
     */
    CoinFeature["STUCK_TRANSACTION_MANAGEMENT_ONCHAIN"] = "stuck-transaction-management-onchain";
})(CoinFeature = exports.CoinFeature || (exports.CoinFeature = {}));
/**
 * Some coins are representations of another underlying asset class. An example
 * is Wrapped Bitcoin, which represents Bitcoin on the Ethereum blockchain.
 *
 * For these coins, the `UnderlyingAsset` provides a link to the actual
 * asset for which the coin is a unit of account.
 */
var UnderlyingAsset;
(function (UnderlyingAsset) {
    UnderlyingAsset["ADA"] = "ada";
    UnderlyingAsset["ALGO"] = "algo";
    UnderlyingAsset["APE"] = "ape";
    UnderlyingAsset["API3"] = "api3";
    UnderlyingAsset["ARBETH"] = "arbeth";
    UnderlyingAsset["BASEETH"] = "baseeth";
    UnderlyingAsset["ATOM"] = "atom";
    UnderlyingAsset["AVAXC"] = "avaxc";
    UnderlyingAsset["AVAXP"] = "avaxp";
    UnderlyingAsset["AXL"] = "AXL";
    UnderlyingAsset["AXLV2"] = "axlv2";
    UnderlyingAsset["BCH"] = "bch";
    UnderlyingAsset["BCHA"] = "bcha";
    UnderlyingAsset["BERA"] = "bera";
    UnderlyingAsset["BLD"] = "bld";
    UnderlyingAsset["BSC"] = "bsc";
    UnderlyingAsset["BSV"] = "bsv";
    UnderlyingAsset["BTC"] = "btc";
    UnderlyingAsset["BTG"] = "btg";
    UnderlyingAsset["DASH"] = "dash";
    UnderlyingAsset["DOT"] = "dot";
    UnderlyingAsset["CELO"] = "celo";
    UnderlyingAsset["COREUM"] = "coreum";
    UnderlyingAsset["CSPR"] = "cspr";
    UnderlyingAsset["ETH"] = "eth";
    UnderlyingAsset["ETH2"] = "eth2";
    UnderlyingAsset["ETHW"] = "ethw";
    UnderlyingAsset["ETC"] = "etc";
    UnderlyingAsset["EOS"] = "eos";
    UnderlyingAsset["ERD"] = "erd";
    UnderlyingAsset["EUR"] = "eur";
    UnderlyingAsset["EURCVV0"] = "eurcvv0";
    UnderlyingAsset["EURCV"] = "eurcv";
    UnderlyingAsset["EUROC"] = "euroc";
    UnderlyingAsset["EURR"] = "eurr";
    UnderlyingAsset["GBP"] = "gbp";
    UnderlyingAsset["GTC"] = "gtc";
    UnderlyingAsset["HASH"] = "hash";
    UnderlyingAsset["HBAR"] = "hbar";
    UnderlyingAsset["INJECTIVE"] = "injective";
    UnderlyingAsset["ISLM"] = "islm";
    UnderlyingAsset["KAVA"] = "kava";
    UnderlyingAsset["LNBTC"] = "lnbtc";
    UnderlyingAsset["LTC"] = "ltc";
    UnderlyingAsset["NEAR"] = "near";
    UnderlyingAsset["OPETH"] = "opeth";
    UnderlyingAsset["OSMO"] = "osmo";
    UnderlyingAsset["POLYGON"] = "polygon";
    UnderlyingAsset["RBTC"] = "rbtc";
    UnderlyingAsset["SEI"] = "sei";
    UnderlyingAsset["SOL"] = "sol";
    UnderlyingAsset["SUI"] = "sui";
    UnderlyingAsset["STX"] = "stx";
    UnderlyingAsset["TIA"] = "tia";
    UnderlyingAsset["TON"] = "ton";
    UnderlyingAsset["TRX"] = "trx";
    UnderlyingAsset["USD"] = "usd";
    UnderlyingAsset["XLM"] = "xlm";
    UnderlyingAsset["XRP"] = "xrp";
    UnderlyingAsset["XTZ"] = "xtz";
    UnderlyingAsset["ZEC"] = "zec";
    UnderlyingAsset["ZETA"] = "zeta";
    UnderlyingAsset["ZKETH"] = "zketh";
    // ERC 20 tokens
    UnderlyingAsset["$Evmosia.com"] = "$evmosia.com";
    UnderlyingAsset["0xREVIEW"] = "0xreview";
    UnderlyingAsset["1INCH"] = "1inch";
    UnderlyingAsset["1UP"] = "1up";
    UnderlyingAsset["3CRV"] = "3crv";
    UnderlyingAsset["AAVE"] = "aave";
    UnderlyingAsset["ABT"] = "abt";
    UnderlyingAsset["ACE"] = "ace";
    UnderlyingAsset["ACEV2"] = "acev2";
    UnderlyingAsset["ACXT"] = "acxt";
    UnderlyingAsset["ACH"] = "ach";
    UnderlyingAsset["ADABEAR"] = "adabear";
    UnderlyingAsset["ADABULL"] = "adabull";
    UnderlyingAsset["AE"] = "ae";
    UnderlyingAsset["AERGO"] = "aergo";
    UnderlyingAsset["AERGO1"] = "aergo1";
    UnderlyingAsset["AGEUR"] = "ageur";
    UnderlyingAsset["AGI"] = "agi";
    UnderlyingAsset["AGIX"] = "agix";
    UnderlyingAsset["AGLD"] = "agld";
    UnderlyingAsset["AGWD"] = "agwd";
    UnderlyingAsset["AION"] = "aion";
    UnderlyingAsset["AJNA"] = "ajna";
    UnderlyingAsset["AKRO"] = "akro";
    UnderlyingAsset["ALCX"] = "alcx";
    UnderlyingAsset["ALD"] = "ald";
    UnderlyingAsset["ALDRIN"] = "aldrin";
    UnderlyingAsset["ALEPH"] = "aleph";
    UnderlyingAsset["ALGOBEAR"] = "algobear";
    UnderlyingAsset["ALGOBULL"] = "algobull";
    UnderlyingAsset["ALGODOOM"] = "algodoom";
    UnderlyingAsset["ALGOHEDGE"] = "algohedge";
    UnderlyingAsset["ALGOMOON"] = "algomoon";
    UnderlyingAsset["ALTDOOM"] = "altdoom";
    UnderlyingAsset["ALTMOON"] = "altmoon";
    UnderlyingAsset["ALI"] = "ali";
    UnderlyingAsset["ALICE"] = "alice";
    UnderlyingAsset["ALK"] = "alk";
    UnderlyingAsset["ALM"] = "alm";
    UnderlyingAsset["ALPHA"] = "alpha";
    UnderlyingAsset["ALTBEAR"] = "altbear";
    UnderlyingAsset["ALTBULL"] = "altbull";
    UnderlyingAsset["ALTHEDGE"] = "althedge";
    UnderlyingAsset["AMKT"] = "amkt";
    UnderlyingAsset["AMN"] = "amn";
    UnderlyingAsset["AMO"] = "amo";
    UnderlyingAsset["AMP"] = "amp";
    UnderlyingAsset["AMPL"] = "ampl";
    UnderlyingAsset["AMON"] = "amon";
    UnderlyingAsset["AMPX"] = "ampx";
    UnderlyingAsset["ANA"] = "ana";
    UnderlyingAsset["ANGLE"] = "angle";
    UnderlyingAsset["ANKR"] = "ankr";
    UnderlyingAsset["ANKRETH"] = "ankreth";
    UnderlyingAsset["ANML"] = "anml";
    UnderlyingAsset["ANT"] = "ant";
    UnderlyingAsset["ANTV2"] = "antv2";
    UnderlyingAsset["AOA"] = "aoa";
    UnderlyingAsset["APPC"] = "appc";
    UnderlyingAsset["APT"] = "apt";
    UnderlyingAsset["AQT"] = "aqt";
    UnderlyingAsset["ARCT"] = "arct";
    UnderlyingAsset["ARCX"] = "arcx";
    UnderlyingAsset["ARMOR"] = "armor";
    UnderlyingAsset["ARTEQ"] = "arteq";
    UnderlyingAsset["ASD"] = "asd";
    UnderlyingAsset["AST"] = "ast";
    UnderlyingAsset["ATA"] = "ata";
    UnderlyingAsset["ATF"] = "atf";
    UnderlyingAsset["ATH"] = "ath";
    UnderlyingAsset["ATL"] = "atl";
    UnderlyingAsset["ATLAS"] = "atlas";
    UnderlyingAsset["ATOMBEAR"] = "atombear";
    UnderlyingAsset["ATOMBULL"] = "atombull";
    UnderlyingAsset["ATRI"] = "atri";
    UnderlyingAsset["AUDD"] = "audd";
    UnderlyingAsset["AUDF"] = "audf";
    UnderlyingAsset["AUDIO"] = "audio";
    UnderlyingAsset["AUDX"] = "audx";
    UnderlyingAsset["AUSD"] = "ausd";
    UnderlyingAsset["AUSDT"] = "ausdt";
    UnderlyingAsset["AUST"] = "aust";
    UnderlyingAsset["AVT"] = "avt";
    UnderlyingAsset["AWBTC"] = "awbtc";
    UnderlyingAsset["AXPR"] = "axpr";
    UnderlyingAsset["AXS"] = "axs";
    UnderlyingAsset["AXSV2"] = "axsv2";
    UnderlyingAsset["AYFI"] = "ayfi";
    UnderlyingAsset["AZUKI"] = "azuki";
    UnderlyingAsset["AZUKI2"] = "azuki2";
    UnderlyingAsset["AZUKIPEPE"] = "azukipepe";
    UnderlyingAsset["BADGER"] = "badger";
    UnderlyingAsset["BAI"] = "bai";
    UnderlyingAsset["BAL"] = "bal";
    UnderlyingAsset["BAND"] = "band";
    UnderlyingAsset["BANK"] = "bank";
    UnderlyingAsset["BAO"] = "bao";
    UnderlyingAsset["BASIC"] = "basic";
    UnderlyingAsset["BAT"] = "bat";
    UnderlyingAsset["BAX"] = "bax";
    UnderlyingAsset["BBANK"] = "bbank";
    UnderlyingAsset["BBSAMO"] = "bbsamo";
    UnderlyingAsset["BBTC"] = "BBTC";
    UnderlyingAsset["BBX"] = "bbx";
    UnderlyingAsset["BCAP"] = "bcap";
    UnderlyingAsset["BCC"] = "bcc";
    UnderlyingAsset["BCHBEAR"] = "bchbear";
    UnderlyingAsset["BCHBULL"] = "bchbull";
    UnderlyingAsset["BCHDOOM"] = "bchdoom";
    UnderlyingAsset["BCHHEDGE"] = "bchhedge";
    UnderlyingAsset["BCHMOON"] = "bchmoon";
    UnderlyingAsset["BCIO"] = "bcio";
    UnderlyingAsset["BCUT"] = "bcut";
    UnderlyingAsset["BCT"] = "bct";
    UnderlyingAsset["BDXN"] = "bdxn";
    UnderlyingAsset["BEAM"] = "beam";
    UnderlyingAsset["BEAR"] = "bear";
    UnderlyingAsset["BEARSHIT"] = "bearshit";
    UnderlyingAsset["BED"] = "bed";
    UnderlyingAsset["BEND"] = "bend";
    UnderlyingAsset["BEPRO"] = "bepro";
    UnderlyingAsset["BGBG"] = "bgbg";
    UnderlyingAsset["BICO"] = "bico";
    UnderlyingAsset["BID"] = "bid";
    UnderlyingAsset["BIDL"] = "bidl";
    UnderlyingAsset["BIRD"] = "bird";
    UnderlyingAsset["BIT"] = "bit";
    UnderlyingAsset["BKX"] = "bkx";
    UnderlyingAsset["BLCT"] = "blct";
    UnderlyingAsset["BLT"] = "blt";
    UnderlyingAsset["BLUR"] = "blur";
    UnderlyingAsset["BLUR0x083"] = "blur0x083";
    UnderlyingAsset["BLUR0xb93"] = "blur0xb93";
    UnderlyingAsset["BLZ"] = "blz";
    UnderlyingAsset["BNB"] = "bnb";
    UnderlyingAsset["BNBBEAR"] = "bnbbear";
    UnderlyingAsset["BNBBULL"] = "bnbbull";
    UnderlyingAsset["BNBDOOM"] = "bnbdoom";
    UnderlyingAsset["BNBHEDGE"] = "bnbhedge";
    UnderlyingAsset["BNBMOON"] = "bnbmoon";
    UnderlyingAsset["BNK"] = "bnk";
    UnderlyingAsset["BNL"] = "bnl";
    UnderlyingAsset["BNT"] = "bnt";
    UnderlyingAsset["BNTY"] = "bnty";
    UnderlyingAsset["BOND"] = "bond";
    UnderlyingAsset["BONE"] = "bone";
    UnderlyingAsset["BORG"] = "borg";
    UnderlyingAsset["BOTTO"] = "botto";
    UnderlyingAsset["BLOCKS"] = "blocks";
    UnderlyingAsset["BOX"] = "box";
    UnderlyingAsset["BOBA"] = "boba";
    UnderlyingAsset["BRD"] = "brd";
    UnderlyingAsset["BRIBE"] = "bribe";
    UnderlyingAsset["BRZ"] = "brz";
    UnderlyingAsset["BSGG"] = "bsgg";
    UnderlyingAsset["BST"] = "bst";
    UnderlyingAsset["BSVBEAR"] = "bsvbear";
    UnderlyingAsset["BSVBULL"] = "bsvbull";
    UnderlyingAsset["BSVDOOM"] = "bsvdoom";
    UnderlyingAsset["BSVHEDGE"] = "bsvhedge";
    UnderlyingAsset["BSVMOON"] = "bsvmoon";
    UnderlyingAsset["BSX"] = "bsx";
    UnderlyingAsset["BTC2XFLI"] = "btc2xfli";
    UnderlyingAsset["BTMXBEAR"] = "btmxbear";
    UnderlyingAsset["BTMXBULL"] = "btmxbull";
    UnderlyingAsset["BTRST"] = "btrst";
    UnderlyingAsset["BTSG"] = "btsg";
    UnderlyingAsset["BTT"] = "btt";
    UnderlyingAsset["BTU"] = "btu";
    UnderlyingAsset["BUIDL"] = "buidl";
    UnderlyingAsset["BULL"] = "bull";
    UnderlyingAsset["BULLSHIT"] = "bullshit";
    UnderlyingAsset["BURP"] = "burp";
    UnderlyingAsset["BUSD"] = "busd";
    UnderlyingAsset["BUY"] = "buy";
    UnderlyingAsset["BPT"] = "bpt";
    UnderlyingAsset["BVOL"] = "bvol";
    UnderlyingAsset["BXX"] = "bxx";
    UnderlyingAsset["BXXV1"] = "bxxv1";
    UnderlyingAsset["BZZ"] = "bzz";
    UnderlyingAsset["C3"] = "c3";
    UnderlyingAsset["C6P"] = "c6p";
    UnderlyingAsset["C8P"] = "c8p";
    UnderlyingAsset["C98"] = "c98";
    UnderlyingAsset["CACXT"] = "cacxt";
    UnderlyingAsset["CADX"] = "cadx";
    UnderlyingAsset["CAG"] = "cag";
    UnderlyingAsset["CANTO"] = "canto";
    UnderlyingAsset["CAPS"] = "caps";
    UnderlyingAsset["CASH"] = "cash";
    UnderlyingAsset["CBAT"] = "cbat";
    UnderlyingAsset["CBC"] = "cbc";
    UnderlyingAsset["CBETH"] = "cbeth";
    UnderlyingAsset["CBRL"] = "cbrl";
    UnderlyingAsset["CCAI"] = "ccai";
    UnderlyingAsset["CCT"] = "cct";
    UnderlyingAsset["CDAG"] = "cdag";
    UnderlyingAsset["CDAI"] = "cdai";
    UnderlyingAsset["CDAIV2"] = "cdaiV2";
    UnderlyingAsset["CDT"] = "cdt";
    UnderlyingAsset["CEL"] = "cel";
    UnderlyingAsset["CELLS"] = "cells";
    UnderlyingAsset["CELR"] = "celr";
    UnderlyingAsset["CERE"] = "cere";
    UnderlyingAsset["CETH"] = "ceth";
    UnderlyingAsset["CFX"] = "cfx";
    UnderlyingAsset["CHAINLINK"] = "chainlink";
    UnderlyingAsset["CHART"] = "chart";
    UnderlyingAsset["CHO"] = "cho";
    UnderlyingAsset["CHFX"] = "chfx";
    UnderlyingAsset["CHR"] = "chr";
    UnderlyingAsset["CHSB"] = "chsb";
    UnderlyingAsset["CHZ"] = "chz";
    UnderlyingAsset["CIBO"] = "cibo";
    UnderlyingAsset["CIX100"] = "cix100";
    UnderlyingAsset["CLIQ"] = "cliq";
    UnderlyingAsset["CLN"] = "cln";
    UnderlyingAsset["CLT"] = "clt";
    UnderlyingAsset["CLXY"] = "clxy";
    UnderlyingAsset["CLV"] = "clv";
    UnderlyingAsset["CMFI"] = "cmfi";
    UnderlyingAsset["CNFI"] = "cnfi";
    UnderlyingAsset["CNG"] = "cng";
    UnderlyingAsset["CNYX"] = "cnyx";
    UnderlyingAsset["COLLAR"] = "collar";
    UnderlyingAsset["COMP"] = "comp";
    UnderlyingAsset["CONV"] = "conv";
    UnderlyingAsset["COPE"] = "cope";
    UnderlyingAsset["CORE"] = "core";
    UnderlyingAsset["COTI"] = "coti";
    UnderlyingAsset["COVAL"] = "coval";
    UnderlyingAsset["COVER"] = "cover";
    UnderlyingAsset["COVERPROTOCOL"] = "coverprotocol";
    UnderlyingAsset["COW"] = "cow";
    UnderlyingAsset["CPAY"] = "cpay";
    UnderlyingAsset["CPLT"] = "cplt";
    UnderlyingAsset["CPOOL"] = "cpool";
    UnderlyingAsset["CQT"] = "cqt";
    UnderlyingAsset["CQX"] = "cqx";
    UnderlyingAsset["CRA"] = "cra";
    UnderlyingAsset["CRDT"] = "crdt";
    UnderlyingAsset["CRE"] = "cre";
    UnderlyingAsset["CREAM"] = "cream";
    UnderlyingAsset["CREP"] = "crep";
    UnderlyingAsset["CRI"] = "cri";
    UnderlyingAsset["CRO"] = "cro";
    UnderlyingAsset["CRV"] = "crv";
    UnderlyingAsset["CRPT"] = "crpt";
    UnderlyingAsset["CRPT1"] = "crpt1";
    UnderlyingAsset["CSLV"] = "cslv";
    UnderlyingAsset["CSOL"] = "csol";
    UnderlyingAsset["CSP"] = "csp";
    UnderlyingAsset["CTSI"] = "ctsi";
    UnderlyingAsset["CTX"] = "ctx";
    UnderlyingAsset["CUBE"] = "cube";
    UnderlyingAsset["CUSD"] = "cusd";
    UnderlyingAsset["CUSDC"] = "cusdc";
    UnderlyingAsset["CVXFXS"] = "cvxfxs";
    UnderlyingAsset["CWAR"] = "cwar";
    UnderlyingAsset["CWBTC"] = "cwbtc";
    UnderlyingAsset["CVC"] = "cvc";
    UnderlyingAsset["CVX"] = "cvx";
    UnderlyingAsset["CXT"] = "cxt";
    UnderlyingAsset["CZRX"] = "czrx";
    UnderlyingAsset["DACXI"] = "dacxi";
    UnderlyingAsset["DADI"] = "dadi";
    UnderlyingAsset["DAMM"] = "damm";
    UnderlyingAsset["DAI"] = "dai";
    UnderlyingAsset["DAO"] = "dao";
    UnderlyingAsset["DAOLANG"] = "daolang";
    UnderlyingAsset["DAR"] = "dar";
    UnderlyingAsset["DATA"] = "data";
    UnderlyingAsset["DATAV2"] = "datav2";
    UnderlyingAsset["DATAECON"] = "dataecon";
    UnderlyingAsset["DAWN"] = "dawn";
    UnderlyingAsset["DEC"] = "dec";
    UnderlyingAsset["DENT"] = "dent";
    UnderlyingAsset["DEP"] = "dep";
    UnderlyingAsset["DEPAY"] = "depay";
    UnderlyingAsset["DEXA"] = "dexa";
    UnderlyingAsset["DFD"] = "dfd";
    UnderlyingAsset["DFI"] = "dfi";
    UnderlyingAsset["DFL"] = "dfl";
    UnderlyingAsset["DFX"] = "dfx";
    UnderlyingAsset["DGCL"] = "dgcl";
    UnderlyingAsset["DGD"] = "dgd";
    UnderlyingAsset["DGLD"] = "dgld";
    UnderlyingAsset["DGX"] = "dgx";
    UnderlyingAsset["DHT"] = "dht";
    UnderlyingAsset["DIGG"] = "digg";
    UnderlyingAsset["DIA"] = "dia";
    UnderlyingAsset["DING"] = "ding";
    UnderlyingAsset["DIPE"] = "dipe";
    UnderlyingAsset["DMG"] = "dmg";
    UnderlyingAsset["DMT"] = "dmt";
    UnderlyingAsset["DNA"] = "dna";
    UnderlyingAsset["DNT"] = "dnt";
    UnderlyingAsset["DODO"] = "dodo";
    UnderlyingAsset["DOG"] = "dog";
    UnderlyingAsset["DOGE"] = "doge";
    UnderlyingAsset["DOGEBEAR"] = "dogebear";
    UnderlyingAsset["DOGEBEAR2021"] = "dogebear2021";
    UnderlyingAsset["DOGEBULL"] = "dogebull";
    UnderlyingAsset["DOOM"] = "doom";
    UnderlyingAsset["DOOMSHIT"] = "doomshit";
    UnderlyingAsset["DOSE"] = "dose";
    UnderlyingAsset["DOTK"] = "dotk";
    UnderlyingAsset["DPAY"] = "dpay";
    UnderlyingAsset["DPI"] = "dpi";
    UnderlyingAsset["DPX"] = "dpx";
    UnderlyingAsset["DPY"] = "dpy";
    UnderlyingAsset["DRAM"] = "dram";
    UnderlyingAsset["DRGNBEAR"] = "drgnbear";
    UnderlyingAsset["DRGNBULL"] = "drgnbull";
    UnderlyingAsset["DRPU"] = "drpu";
    UnderlyingAsset["DRV"] = "drv";
    UnderlyingAsset["DUC"] = "duc";
    UnderlyingAsset["DUCK"] = "duck";
    UnderlyingAsset["DUSD"] = "dusd";
    UnderlyingAsset["DUST"] = "dust";
    UnderlyingAsset["DX1U"] = "dx1u";
    UnderlyingAsset["DXGT"] = "dxgt";
    UnderlyingAsset["DXO"] = "dxo";
    UnderlyingAsset["DXPT"] = "dxpt";
    UnderlyingAsset["DXST"] = "dxst";
    UnderlyingAsset["DYDX"] = "dydx";
    UnderlyingAsset["DYN"] = "dyn";
    UnderlyingAsset["EASY"] = "easy";
    UnderlyingAsset["EBTCQ"] = "ebtcq";
    UnderlyingAsset["ECHT"] = "echt";
    UnderlyingAsset["ECOX"] = "ecox";
    UnderlyingAsset["EDEN"] = "eden";
    UnderlyingAsset["EDISON"] = "edison";
    UnderlyingAsset["EDLC"] = "edlc";
    UnderlyingAsset["EDO"] = "edo";
    UnderlyingAsset["ELON"] = "elon";
    UnderlyingAsset["EMB"] = "emb";
    UnderlyingAsset["EDN"] = "edn";
    UnderlyingAsset["EDR"] = "edr";
    UnderlyingAsset["EFI"] = "efi";
    UnderlyingAsset["EGL"] = "egl";
    UnderlyingAsset["EGLD"] = "egld";
    UnderlyingAsset["EGOLD"] = "egold";
    UnderlyingAsset["EIGEN"] = "eigen";
    UnderlyingAsset["ELF"] = "elf";
    UnderlyingAsset["ELU"] = "elu";
    UnderlyingAsset["EMAID"] = "emaid";
    UnderlyingAsset["EMX"] = "emx";
    UnderlyingAsset["ENA"] = "ena";
    UnderlyingAsset["ENG"] = "eng";
    UnderlyingAsset["ENJ"] = "enj";
    UnderlyingAsset["ENS"] = "ens";
    UnderlyingAsset["EON"] = "eon";
    UnderlyingAsset["EOP"] = "eop";
    UnderlyingAsset["EOSBEAR"] = "eosbear";
    UnderlyingAsset["EOSBULL"] = "eosbull";
    UnderlyingAsset["EOSDOOM"] = "eosdoom";
    UnderlyingAsset["EOSHEDGE"] = "eoshedge";
    UnderlyingAsset["EOSMOON"] = "eosmoon";
    UnderlyingAsset["EQO"] = "eqo";
    UnderlyingAsset["ETA"] = "eta";
    UnderlyingAsset["ETHBULL"] = "ethbull";
    UnderlyingAsset["ETCBEAR"] = "etcbear";
    UnderlyingAsset["ETCBULL"] = "etcbull";
    UnderlyingAsset["ETCDOOM"] = "etcdoom";
    UnderlyingAsset["ETCHEDOOM"] = "etchedoom";
    UnderlyingAsset["ETCMOON"] = "etcmoon";
    UnderlyingAsset["ETHBEAR"] = "ethbear";
    UnderlyingAsset["ETHDOOM"] = "ethdoom";
    UnderlyingAsset["ETHFI"] = "ethfi";
    UnderlyingAsset["ETHHEDGE"] = "ethhedge";
    UnderlyingAsset["ETHMOON"] = "ethmoon";
    UnderlyingAsset["ETHOPT"] = "ethopt";
    UnderlyingAsset["ETHOS"] = "ethos";
    UnderlyingAsset["ETHX"] = "ethx";
    UnderlyingAsset["ETV"] = "etv";
    UnderlyingAsset["ETX"] = "etx";
    UnderlyingAsset["EUL"] = "eul";
    UnderlyingAsset["EURE"] = "eure";
    UnderlyingAsset["EURL"] = "eurl";
    UnderlyingAsset["EUROE"] = "euroe";
    UnderlyingAsset["EURS"] = "eurs";
    UnderlyingAsset["EURST"] = "eurst";
    UnderlyingAsset["EURT"] = "eurt";
    UnderlyingAsset["EURX"] = "eurx";
    UnderlyingAsset["EUX"] = "eux";
    UnderlyingAsset["EVER"] = "ever";
    UnderlyingAsset["EVERY"] = "every";
    UnderlyingAsset["EVRY"] = "evry";
    UnderlyingAsset["EVX"] = "evx";
    UnderlyingAsset["EXCHBEAR"] = "exchbear";
    UnderlyingAsset["EXCHBULL"] = "exchbull";
    UnderlyingAsset["EXCHDOOM"] = "exchdoom";
    UnderlyingAsset["EXCHHEDGE"] = "exchhedge";
    UnderlyingAsset["EXCHMOON"] = "exchmoon";
    UnderlyingAsset["EXE"] = "exe";
    UnderlyingAsset["FANT"] = "fant";
    UnderlyingAsset["FARM"] = "farm";
    UnderlyingAsset["FEI"] = "fei";
    UnderlyingAsset["FET"] = "fet";
    UnderlyingAsset["FET1"] = "fet1";
    UnderlyingAsset["FDT"] = "fdt";
    UnderlyingAsset["FF1"] = "ff1";
    UnderlyingAsset["FF6000"] = "ff6000";
    UnderlyingAsset["FFT"] = "fft";
    UnderlyingAsset["FIDA"] = "fida";
    UnderlyingAsset["FIDU"] = "fidu";
    UnderlyingAsset["FIN"] = "fin";
    UnderlyingAsset["FIRE"] = "fire";
    UnderlyingAsset["FIRSTBLOOD"] = "firstblood";
    UnderlyingAsset["FIXED"] = "fixed";
    UnderlyingAsset["FLIP"] = "flip";
    UnderlyingAsset["FLOKI"] = "floki";
    UnderlyingAsset["FLY"] = "fly";
    UnderlyingAsset["FMF"] = "fmf";
    UnderlyingAsset["FOREX"] = "forex";
    UnderlyingAsset["FORT"] = "fort";
    UnderlyingAsset["FORTH"] = "forth";
    UnderlyingAsset["FOX"] = "fox";
    UnderlyingAsset["FPIS"] = "fpis";
    UnderlyingAsset["FRAX"] = "frax";
    UnderlyingAsset["FRONT"] = "front";
    UnderlyingAsset["FT"] = "ft";
    UnderlyingAsset["FTM"] = "ftm";
    UnderlyingAsset["FTT"] = "ftt";
    UnderlyingAsset["FTT20"] = "ftt20";
    UnderlyingAsset["FTX2"] = "ftx2";
    UnderlyingAsset["FUCKFTX"] = "fuckftx";
    UnderlyingAsset["FUN"] = "fun";
    UnderlyingAsset["FWB"] = "fwb";
    UnderlyingAsset["FX"] = "fx";
    UnderlyingAsset["FXRT"] = "fxrt";
    UnderlyingAsset["FXS"] = "fxs";
    UnderlyingAsset["G"] = "g";
    UnderlyingAsset["GAL"] = "gal";
    UnderlyingAsset["GALA"] = "gala";
    UnderlyingAsset["GALAV2"] = "galav2";
    UnderlyingAsset["GAME.COM"] = "game.com";
    UnderlyingAsset["GAMMA"] = "gamma";
    UnderlyingAsset["sol:gari"] = "sol:gari";
    UnderlyingAsset["GAS"] = "gas";
    UnderlyingAsset["GATE"] = "gate";
    UnderlyingAsset["GBPT"] = "gbpt";
    UnderlyingAsset["GBPX"] = "gbpx";
    UnderlyingAsset["GDT"] = "gdt";
    UnderlyingAsset["GEAR"] = "gear";
    UnderlyingAsset["GEC"] = "gec";
    UnderlyingAsset["GEL"] = "gel";
    UnderlyingAsset["GEN"] = "gen";
    UnderlyingAsset["GENE"] = "gene";
    UnderlyingAsset["GENIE"] = "genie";
    UnderlyingAsset["GF"] = "gf";
    UnderlyingAsset["GFI"] = "gfi";
    UnderlyingAsset["GHUB"] = "ghub";
    UnderlyingAsset["GIGDROP"] = "gigdrop";
    UnderlyingAsset["GIV"] = "giv";
    UnderlyingAsset["GLDX"] = "gldx";
    UnderlyingAsset["GLM"] = "glm";
    UnderlyingAsset["GMT"] = "gmt";
    UnderlyingAsset["sol:gmt"] = "sol:gmt";
    UnderlyingAsset["GNO"] = "gno";
    UnderlyingAsset["GNT"] = "gnt";
    UnderlyingAsset["GODS"] = "gods";
    UnderlyingAsset["GOHM"] = "gohm";
    UnderlyingAsset["GOG"] = "gog";
    UnderlyingAsset["GOLD"] = "gold";
    UnderlyingAsset["GOM"] = "gom";
    UnderlyingAsset["GOMINING"] = "gomining";
    UnderlyingAsset["GOT"] = "got";
    UnderlyingAsset["GRID"] = "grid";
    UnderlyingAsset["GRT"] = "grt";
    UnderlyingAsset["GST"] = "gst";
    UnderlyingAsset["GT"] = "gt";
    UnderlyingAsset["GTAAVE18DP"] = "gtaave18dp";
    UnderlyingAsset["GTBAT18DP"] = "gtbat18dp";
    UnderlyingAsset["GTCOMP18DP"] = "gtcomp18dp";
    UnderlyingAsset["GTGRT18DP"] = "gtgrt18dp";
    UnderlyingAsset["GTLINK18DP"] = "gtlink18dp";
    UnderlyingAsset["GTMKR18DP"] = "gtmkr18dp";
    UnderlyingAsset["GTSNX18DP"] = "gtsnx18dp";
    UnderlyingAsset["GTUNI18DP"] = "gtuni18dp";
    UnderlyingAsset["GTUSDT6DP"] = "gtusdt6dp";
    UnderlyingAsset["GTYFI18DP"] = "gtyfi18dp";
    UnderlyingAsset["GTWBTC8DP"] = "gtwbtc8dp";
    UnderlyingAsset["GTO"] = "gto";
    UnderlyingAsset["GTERC2DP"] = "gterc2dp";
    UnderlyingAsset["GTERC6DP"] = "gterc6dp";
    UnderlyingAsset["GTERC18DP"] = "gterc18dp";
    UnderlyingAsset["GUSD"] = "gusd";
    UnderlyingAsset["GUSDT"] = "gusdt";
    UnderlyingAsset["GXC"] = "gxc";
    UnderlyingAsset["GXT"] = "gxt";
    UnderlyingAsset["GYEN"] = "gyen";
    UnderlyingAsset["HBB"] = "hbb";
    UnderlyingAsset["HBTC"] = "hbtc";
    UnderlyingAsset["HCN"] = "hcn";
    UnderlyingAsset["HDO"] = "hdo";
    UnderlyingAsset["HEDG"] = "hedg";
    UnderlyingAsset["HEDGE"] = "hedge";
    UnderlyingAsset["HEDGESHIT"] = "hedgeshit";
    UnderlyingAsset["HEX"] = "hex";
    UnderlyingAsset["HFT"] = "hft";
    UnderlyingAsset["HGET"] = "hget";
    UnderlyingAsset["HIT"] = "hit";
    UnderlyingAsset["HKDX"] = "hkdx";
    UnderlyingAsset["HLC"] = "hlc";
    UnderlyingAsset["HMT"] = "hmt";
    UnderlyingAsset["sol:hnt"] = "sol:hnt";
    UnderlyingAsset["HOLD"] = "hold";
    UnderlyingAsset["HOLY"] = "holy";
    UnderlyingAsset["HOP"] = "hop";
    UnderlyingAsset["HOT"] = "hot";
    UnderlyingAsset["HPO"] = "hpo";
    UnderlyingAsset["HQG"] = "hqg";
    UnderlyingAsset["HQT"] = "hqt";
    UnderlyingAsset["HST"] = "hst";
    UnderlyingAsset["HT"] = "ht";
    UnderlyingAsset["HTBEAR"] = "htbear";
    UnderlyingAsset["HTBULL"] = "htbull";
    UnderlyingAsset["HTDOOM"] = "htdoom";
    UnderlyingAsset["HTHEDGE"] = "hthedge";
    UnderlyingAsset["HTMOON"] = "htmoon";
    UnderlyingAsset["HUM"] = "hum";
    UnderlyingAsset["HUMV2"] = "humv2";
    UnderlyingAsset["HUSD"] = "husd";
    UnderlyingAsset["HXRO"] = "hxro";
    UnderlyingAsset["HYB"] = "hyb";
    UnderlyingAsset["HYDRO"] = "hydro";
    UnderlyingAsset["HYDROPROTOCOL"] = "hydroprotocol";
    UnderlyingAsset["I8"] = "i8";
    UnderlyingAsset["IBEUR"] = "ibeur";
    UnderlyingAsset["IBOX"] = "ibox";
    UnderlyingAsset["IBVOL"] = "ibvol";
    UnderlyingAsset["ICETH"] = "iceth";
    UnderlyingAsset["IDEX"] = "idex";
    UnderlyingAsset["IDRC"] = "idrc";
    UnderlyingAsset["IDRT"] = "idrt";
    UnderlyingAsset["ILV"] = "ilv";
    UnderlyingAsset["IMX"] = "imx";
    UnderlyingAsset["IMXV2"] = "imxv2";
    UnderlyingAsset["INCX"] = "incx";
    UnderlyingAsset["IND"] = "ind";
    UnderlyingAsset["INDEX"] = "index";
    UnderlyingAsset["INDI"] = "indi";
    UnderlyingAsset["INF"] = "inf";
    UnderlyingAsset["INJ"] = "inj";
    UnderlyingAsset["INJV2"] = "injv2";
    UnderlyingAsset["INST"] = "inst";
    UnderlyingAsset["INSUR"] = "insur";
    UnderlyingAsset["INV"] = "inv";
    UnderlyingAsset["INX"] = "inx";
    UnderlyingAsset["IOTX"] = "iotx";
    UnderlyingAsset["IP3"] = "ip3";
    UnderlyingAsset["ISF"] = "isf";
    UnderlyingAsset["ISR"] = "isr";
    UnderlyingAsset["IVO"] = "ivo";
    UnderlyingAsset["IVY"] = "ivy";
    UnderlyingAsset["JASMY"] = "jasmy";
    UnderlyingAsset["JBC"] = "jbc";
    UnderlyingAsset["JCR"] = "jcr";
    UnderlyingAsset["JCG"] = "jcg";
    UnderlyingAsset["sol:jet"] = "sol:jet";
    UnderlyingAsset["JFIN"] = "jfin";
    UnderlyingAsset["JPYX"] = "jpyx";
    UnderlyingAsset["JSOL"] = "jsol";
    UnderlyingAsset["KARATE"] = "karate";
    UnderlyingAsset["KARMA"] = "karma";
    UnderlyingAsset["KAS"] = "kas";
    UnderlyingAsset["KCASH"] = "kcash";
    UnderlyingAsset["KEEP"] = "keep";
    UnderlyingAsset["KEY"] = "key";
    UnderlyingAsset["KILL0"] = "kill0";
    UnderlyingAsset["KIN"] = "kin";
    UnderlyingAsset["sol:kin"] = "sol:kin";
    UnderlyingAsset["KINE"] = "kine";
    UnderlyingAsset["KING"] = "king";
    UnderlyingAsset["KINTO"] = "kinto";
    UnderlyingAsset["KIRO"] = "kiro";
    UnderlyingAsset["KISHUI"] = "kishui";
    UnderlyingAsset["KITTY"] = "kitty";
    UnderlyingAsset["KNC"] = "knc";
    UnderlyingAsset["KNC2"] = "knc2";
    UnderlyingAsset["KOIN"] = "koin";
    UnderlyingAsset["KOL"] = "kol";
    UnderlyingAsset["KOZ"] = "koz";
    UnderlyingAsset["KP3R"] = "kp3r";
    UnderlyingAsset["KROM"] = "krom";
    UnderlyingAsset["KTRC"] = "ktrc";
    UnderlyingAsset["KZE"] = "kze";
    UnderlyingAsset["L3USD"] = "l3usd";
    UnderlyingAsset["LA"] = "la";
    UnderlyingAsset["LADYS"] = "ladys";
    UnderlyingAsset["LAYER"] = "layer";
    UnderlyingAsset["LAYERZERO"] = "layerzero";
    UnderlyingAsset["LBA"] = "lba";
    UnderlyingAsset["LCX"] = "lcx";
    UnderlyingAsset["LDO"] = "ldo";
    UnderlyingAsset["LEND"] = "lend";
    UnderlyingAsset["LEO"] = "leo";
    UnderlyingAsset["LEOBEAR"] = "leobear";
    UnderlyingAsset["LEOBULL"] = "leobull";
    UnderlyingAsset["LEODOOM"] = "leodoom";
    UnderlyingAsset["LEOHEDGE"] = "leohedge";
    UnderlyingAsset["LEOMOON"] = "leomoon";
    UnderlyingAsset["LEV"] = "lev";
    UnderlyingAsset["LGO"] = "lgo";
    UnderlyingAsset["LIEN"] = "lien";
    UnderlyingAsset["LIF3"] = "lif3";
    UnderlyingAsset["LIKE"] = "like";
    UnderlyingAsset["LINA"] = "lina";
    UnderlyingAsset["LINK"] = "link";
    UnderlyingAsset["LINKBEAR"] = "linkbear";
    UnderlyingAsset["LINKBULL"] = "linkbull";
    UnderlyingAsset["LION"] = "lion";
    UnderlyingAsset["LIT"] = "lit";
    UnderlyingAsset["LITH"] = "lith";
    UnderlyingAsset["LITv2"] = "litv2";
    UnderlyingAsset["LKR"] = "lkr";
    UnderlyingAsset["LMWR"] = "lmwr";
    UnderlyingAsset["LNC"] = "lnc";
    UnderlyingAsset["LOOKS"] = "looks";
    UnderlyingAsset["LOOM"] = "loom";
    UnderlyingAsset["LOOM1"] = "loom1";
    UnderlyingAsset["LOVE"] = "love";
    UnderlyingAsset["LOWB"] = "lowb";
    UnderlyingAsset["LPT"] = "lpt";
    UnderlyingAsset["LQID"] = "lqid";
    UnderlyingAsset["LQTY"] = "lqty";
    UnderlyingAsset["LRC"] = "lrc";
    UnderlyingAsset["LRCV2"] = "lrcv2";
    UnderlyingAsset["LSETH"] = "lseth";
    UnderlyingAsset["LTCBEAR"] = "ltcbear";
    UnderlyingAsset["LTCBULL"] = "ltcbull";
    UnderlyingAsset["LTCDOOM"] = "ltcdoom";
    UnderlyingAsset["LTCHEDGE"] = "ltchedge";
    UnderlyingAsset["LTCMOON"] = "ltcmoon";
    UnderlyingAsset["LUA"] = "lua";
    UnderlyingAsset["LUNA"] = "luna";
    UnderlyingAsset["LUNAWORMHOLE"] = "lunawormhole";
    UnderlyingAsset["LYN"] = "lyn";
    UnderlyingAsset["LYXE"] = "lyxe";
    UnderlyingAsset["MAGIC"] = "magic";
    UnderlyingAsset["MANA"] = "mana";
    UnderlyingAsset["MANTRADAO"] = "mantradao";
    UnderlyingAsset["MAPS"] = "maps";
    UnderlyingAsset["MASA"] = "masa";
    UnderlyingAsset["MASK"] = "mask";
    UnderlyingAsset["MATH"] = "math";
    UnderlyingAsset["MATIC"] = "matic";
    UnderlyingAsset["MATICBEAR"] = "maticbear";
    UnderlyingAsset["MATICBEAR2021"] = "maticbear2021";
    UnderlyingAsset["MATICBULL"] = "maticbull";
    UnderlyingAsset["MATTER"] = "matter";
    UnderlyingAsset["MAV"] = "mav";
    UnderlyingAsset["MBS"] = "mbs";
    UnderlyingAsset["MCAU"] = "mcau";
    UnderlyingAsset["MCB"] = "mcb";
    UnderlyingAsset["MCDAI"] = "mcdai";
    UnderlyingAsset["MCO"] = "mco";
    UnderlyingAsset["MCO2"] = "mco2";
    UnderlyingAsset["MCS"] = "mcs";
    UnderlyingAsset["MCX"] = "mcx";
    UnderlyingAsset["MDFC"] = "mdfc";
    UnderlyingAsset["MDT"] = "mdt";
    UnderlyingAsset["MDX"] = "mdx";
    UnderlyingAsset["MEAN"] = "mean";
    UnderlyingAsset["MEDIA"] = "media";
    UnderlyingAsset["MEDIAv2"] = "mediav2";
    UnderlyingAsset["MEDX"] = "medx";
    UnderlyingAsset["MEME"] = "meme";
    UnderlyingAsset["MEOW"] = "meow";
    UnderlyingAsset["MER"] = "mer";
    UnderlyingAsset["MET"] = "met";
    UnderlyingAsset["META"] = "meta";
    UnderlyingAsset["METIS"] = "metis";
    UnderlyingAsset["MEW"] = "mew";
    UnderlyingAsset["MFG"] = "mfg";
    UnderlyingAsset["MFPH"] = "mfph";
    UnderlyingAsset["MFT"] = "mft";
    UnderlyingAsset["MIDBEAR"] = "midbear";
    UnderlyingAsset["MIDBULL"] = "midbull";
    UnderlyingAsset["MIDDOOM"] = "middoom";
    UnderlyingAsset["MIDHEDGE"] = "midhedge";
    UnderlyingAsset["MIDMOON"] = "midmoon";
    UnderlyingAsset["MILKV2"] = "milkv2";
    UnderlyingAsset["MIM"] = "mim";
    UnderlyingAsset["MIR"] = "mir";
    UnderlyingAsset["MITH"] = "mith";
    UnderlyingAsset["MIX"] = "mix";
    UnderlyingAsset["MIZN"] = "mizn";
    UnderlyingAsset["MKR"] = "mkr";
    UnderlyingAsset["MLN"] = "mln";
    UnderlyingAsset["MNS"] = "mns";
    UnderlyingAsset["MNT"] = "mnt";
    UnderlyingAsset["MNDE"] = "mnde";
    UnderlyingAsset["MOC"] = "moc";
    UnderlyingAsset["MOCA"] = "moca";
    UnderlyingAsset["MOCHI"] = "mochi";
    UnderlyingAsset["MOF"] = "mof";
    UnderlyingAsset["MOG"] = "mog";
    UnderlyingAsset["MOH"] = "moh";
    UnderlyingAsset["MOON"] = "moon";
    UnderlyingAsset["MOONSHIT"] = "moonshit";
    UnderlyingAsset["MOTHER"] = "mother";
    UnderlyingAsset["MNGO"] = "mngo";
    UnderlyingAsset["MPAY"] = "mpay";
    UnderlyingAsset["MPL"] = "mpl";
    UnderlyingAsset["sol:mplx"] = "sol:mplx";
    UnderlyingAsset["MRTWEET"] = "mrtweet";
    UnderlyingAsset["MSN"] = "msn";
    UnderlyingAsset["MSOL"] = "msol";
    UnderlyingAsset["MTA"] = "mta";
    UnderlyingAsset["MTCN"] = "mtcn";
    UnderlyingAsset["MTH"] = "mth";
    UnderlyingAsset["MTL"] = "mtl";
    UnderlyingAsset["MTV"] = "mtv";
    UnderlyingAsset["MUSD"] = "musd";
    UnderlyingAsset["MVL"] = "mvl";
    UnderlyingAsset["MVI"] = "mvi";
    UnderlyingAsset["MWT"] = "mwt";
    UnderlyingAsset["MYRC"] = "myrc";
    UnderlyingAsset["MYTH"] = "myth";
    UnderlyingAsset["NAAI"] = "naai";
    UnderlyingAsset["NAS"] = "nas";
    UnderlyingAsset["NCT"] = "nct";
    UnderlyingAsset["NDX"] = "ndx";
    UnderlyingAsset["NEAR-ERC20"] = "near-erc20";
    UnderlyingAsset["NEU"] = "neu";
    UnderlyingAsset["NEWO"] = "newo";
    UnderlyingAsset["NEXO"] = "nexo";
    UnderlyingAsset["NFCWIN-SB-2021"] = "nfcwin-sb-2021";
    UnderlyingAsset["NFTX"] = "nftx";
    UnderlyingAsset["NGNT"] = "ngnt";
    UnderlyingAsset["NIAX"] = "niax";
    UnderlyingAsset["NKN"] = "nkn";
    UnderlyingAsset["NMR"] = "nmr";
    UnderlyingAsset["NOSANA"] = "nosana";
    UnderlyingAsset["NOTE"] = "note";
    UnderlyingAsset["NOVA"] = "nova";
    UnderlyingAsset["NPXS"] = "npxs";
    UnderlyingAsset["NS2DRP"] = "ns2drp";
    UnderlyingAsset["NU"] = "nu";
    UnderlyingAsset["NULS"] = "nuls";
    UnderlyingAsset["NUTS"] = "nuts";
    UnderlyingAsset["NYM"] = "nym";
    UnderlyingAsset["NZDX"] = "nzdx";
    UnderlyingAsset["OAX"] = "oax";
    UnderlyingAsset["OCEAN"] = "ocean";
    UnderlyingAsset["OCEANV2"] = "oceanv2";
    UnderlyingAsset["OCTAV"] = "octav";
    UnderlyingAsset["OGN"] = "ogn";
    UnderlyingAsset["OGV"] = "ogv";
    UnderlyingAsset["OKB"] = "okb";
    UnderlyingAsset["OKBBEAR"] = "okbbear";
    UnderlyingAsset["OKBBULL"] = "okbbull";
    UnderlyingAsset["OKBDOOM"] = "okbdoom";
    UnderlyingAsset["OKBHEDGE"] = "okbhedge";
    UnderlyingAsset["OKBMOON"] = "okbmoon";
    UnderlyingAsset["OM"] = "om";
    UnderlyingAsset["OMG"] = "omg";
    UnderlyingAsset["OMNIA"] = "omnia";
    UnderlyingAsset["ONDO"] = "ondo";
    UnderlyingAsset["ONL"] = "onl";
    UnderlyingAsset["OP"] = "op";
    UnderlyingAsset["OPIUM"] = "opium";
    UnderlyingAsset["OPT"] = "opt";
    UnderlyingAsset["ORAI"] = "orai";
    UnderlyingAsset["ORBS"] = "orbs";
    UnderlyingAsset["ORC"] = "orc";
    UnderlyingAsset["sol:orca"] = "sol:orca";
    UnderlyingAsset["OS"] = "os";
    UnderlyingAsset["OSETH"] = "oseth";
    UnderlyingAsset["OUSD"] = "ousd";
    UnderlyingAsset["OUSG"] = "ousg";
    UnderlyingAsset["OWN"] = "own";
    UnderlyingAsset["OXT"] = "oxt";
    UnderlyingAsset["OXY"] = "oxy";
    UnderlyingAsset["OHM"] = "ohm";
    UnderlyingAsset["PACT"] = "pact";
    UnderlyingAsset["PAI"] = "pai";
    UnderlyingAsset["PAR"] = "par";
    UnderlyingAsset["PASS"] = "pass";
    UnderlyingAsset["PAU"] = "pau";
    UnderlyingAsset["PAX"] = "pax";
    UnderlyingAsset["PAXG"] = "paxg";
    UnderlyingAsset["PAXGBEAR"] = "paxgbear";
    UnderlyingAsset["PAXGBULL"] = "paxgbull";
    UnderlyingAsset["PAY"] = "pay";
    UnderlyingAsset["PBCH"] = "pbch";
    UnderlyingAsset["PBTC"] = "pbtc";
    UnderlyingAsset["PDATA"] = "pdata";
    UnderlyingAsset["PDI"] = "pdi";
    UnderlyingAsset["PEAQ"] = "peaq";
    UnderlyingAsset["PEBBLE"] = "pebble";
    UnderlyingAsset["PEG"] = "peg";
    UnderlyingAsset["PENDLE"] = "pendle";
    UnderlyingAsset["PEOPLE"] = "people";
    UnderlyingAsset["PEPE"] = "pepe";
    UnderlyingAsset["PERP"] = "perp";
    UnderlyingAsset["PETH"] = "peth";
    UnderlyingAsset["PHA"] = "pha";
    UnderlyingAsset["PHNX"] = "phnx";
    UnderlyingAsset["PICK"] = "pick";
    UnderlyingAsset["PICKLE"] = "pickle";
    UnderlyingAsset["PIE"] = "pie";
    UnderlyingAsset["PINE"] = "pine";
    UnderlyingAsset["PIRATE"] = "pirate";
    UnderlyingAsset["PLC"] = "plc";
    UnderlyingAsset["PFCT"] = "pfct";
    UnderlyingAsset["PLANET"] = "planet";
    UnderlyingAsset["PLNX"] = "plnx";
    UnderlyingAsset["PLX"] = "plx";
    UnderlyingAsset["PMA"] = "pma";
    UnderlyingAsset["POL"] = "pol";
    UnderlyingAsset["POLIS"] = "polis";
    UnderlyingAsset["POLY"] = "poly";
    UnderlyingAsset["POLS"] = "pols";
    UnderlyingAsset["PONYS"] = "ponys";
    UnderlyingAsset["PORT"] = "port";
    UnderlyingAsset["POWR"] = "powr";
    UnderlyingAsset["PPT"] = "ppt";
    UnderlyingAsset["PRDX"] = "prdx";
    UnderlyingAsset["PRINTS"] = "prints";
    UnderlyingAsset["PRISM"] = "prism";
    UnderlyingAsset["PRO"] = "pro";
    UnderlyingAsset["PROM"] = "prom";
    UnderlyingAsset["PROS"] = "pros";
    UnderlyingAsset["PRT"] = "prt";
    UnderlyingAsset["PRTS"] = "prts";
    UnderlyingAsset["PSOL"] = "psol";
    UnderlyingAsset["PSP"] = "psp";
    UnderlyingAsset["PSTAKE"] = "pstake";
    UnderlyingAsset["PSY"] = "psy";
    UnderlyingAsset["PTU"] = "ptu";
    UnderlyingAsset["PUNDIX"] = "pundix";
    UnderlyingAsset["PUSD"] = "pusd";
    UnderlyingAsset["PXP"] = "pxp";
    UnderlyingAsset["PYR"] = "pyr";
    UnderlyingAsset["PYUSD"] = "pyusd";
    UnderlyingAsset["QASH"] = "qash";
    UnderlyingAsset["QCAD"] = "qcad";
    UnderlyingAsset["sol:qcad"] = "sol:qcad";
    UnderlyingAsset["QOM"] = "qom";
    UnderlyingAsset["QUICK"] = "quick";
    UnderlyingAsset["QDT"] = "qdt";
    UnderlyingAsset["QKC"] = "qkc";
    UnderlyingAsset["QLINDO"] = "qlindo";
    UnderlyingAsset["QNT"] = "qnt";
    UnderlyingAsset["QRDO"] = "qrdo";
    UnderlyingAsset["QRL"] = "qrl";
    UnderlyingAsset["QSP"] = "qsp";
    UnderlyingAsset["QVT"] = "qvt";
    UnderlyingAsset["RAD"] = "rad";
    UnderlyingAsset["RADAR"] = "radar";
    UnderlyingAsset["RAIN"] = "rain";
    UnderlyingAsset["RALPH"] = "ralph";
    UnderlyingAsset["RAMP"] = "ramp";
    UnderlyingAsset["RARE"] = "rare";
    UnderlyingAsset["RARI"] = "rari";
    UnderlyingAsset["RAY"] = "ray";
    UnderlyingAsset["sol:ray"] = "sol:ray";
    UnderlyingAsset["RAZOR"] = "razor";
    UnderlyingAsset["RBANK"] = "rbank";
    UnderlyingAsset["RBN"] = "rbn";
    UnderlyingAsset["RBX"] = "rbx";
    UnderlyingAsset["RBY"] = "rby";
    UnderlyingAsset["RCOIN"] = "rcoin";
    UnderlyingAsset["RCT"] = "rct";
    UnderlyingAsset["RDN"] = "rdn";
    UnderlyingAsset["REAL"] = "real";
    UnderlyingAsset["REB"] = "reb";
    UnderlyingAsset["REBL"] = "rebl";
    UnderlyingAsset["REEF"] = "reef";
    UnderlyingAsset["REF"] = "ref";
    UnderlyingAsset["REKT"] = "rekt";
    UnderlyingAsset["REKTGAME"] = "rektgame";
    UnderlyingAsset["REN"] = "ren";
    UnderlyingAsset["RENBTC"] = "renbtc";
    UnderlyingAsset["RENDOGE"] = "rendoge";
    UnderlyingAsset["REP"] = "rep";
    UnderlyingAsset["REPV2"] = "repv2";
    UnderlyingAsset["REQ"] = "REQ";
    UnderlyingAsset["RETH-ROCKET"] = "reth-rocket";
    UnderlyingAsset["RETH-STAFI"] = "reth-stafi";
    UnderlyingAsset["RETH-H"] = "reth-h";
    UnderlyingAsset["RETH2"] = "reth2";
    UnderlyingAsset["REVV"] = "revv";
    UnderlyingAsset["RFOX"] = "rfox";
    UnderlyingAsset["RFR"] = "rfr";
    UnderlyingAsset["RFUEL"] = "rfuel";
    UnderlyingAsset["RGT"] = "rgt";
    UnderlyingAsset["RIF"] = "rif";
    UnderlyingAsset["RINGX"] = "ringx";
    UnderlyingAsset["RIO"] = "rio";
    UnderlyingAsset["RLC"] = "rlc";
    UnderlyingAsset["RLY"] = "rly";
    UnderlyingAsset["RN"] = "rn";
    UnderlyingAsset["RND"] = "rnd";
    UnderlyingAsset["RNDR"] = "rndr";
    UnderlyingAsset["ROOK"] = "rook";
    UnderlyingAsset["RON"] = "ron";
    UnderlyingAsset["RONC"] = "ronc";
    UnderlyingAsset["ROOBEE"] = "roobee";
    UnderlyingAsset["RPL"] = "rpl";
    UnderlyingAsset["RSR"] = "rsr";
    UnderlyingAsset["RSWETH"] = "rsweth";
    UnderlyingAsset["RUBX"] = "rubx";
    UnderlyingAsset["RUEDATK"] = "ruedatk";
    UnderlyingAsset["RUN"] = "run";
    UnderlyingAsset["RUNE"] = "rune";
    UnderlyingAsset["RVR"] = "rvr";
    UnderlyingAsset["RYOSHI"] = "ryoshi";
    UnderlyingAsset["SAIL"] = "sail";
    UnderlyingAsset["SAITABIT"] = "saitabit";
    UnderlyingAsset["SALT"] = "salt";
    UnderlyingAsset["SAND"] = "sand";
    UnderlyingAsset["SASHIMI"] = "sashimi";
    UnderlyingAsset["SAMO"] = "samo";
    UnderlyingAsset["SBC"] = "sbc";
    UnderlyingAsset["sol:sbc"] = "sol:sbc";
    UnderlyingAsset["SBF"] = "sbf";
    UnderlyingAsset["SBR"] = "sbr";
    // Saber IOU Token (Liquidity Mining Rewards)
    UnderlyingAsset["SBRIOU"] = "sbriou";
    UnderlyingAsset["SCNSOL"] = "scnsol";
    UnderlyingAsset["SCOPE"] = "scope";
    UnderlyingAsset["SD"] = "sd";
    UnderlyingAsset["SDL"] = "sdl";
    UnderlyingAsset["SECO"] = "seco";
    UnderlyingAsset["SETH"] = "seth";
    UnderlyingAsset["SETH-H"] = "seth-h";
    UnderlyingAsset["SETH2"] = "seth2";
    UnderlyingAsset["SEWERCOIN"] = "sewercoin";
    UnderlyingAsset["SFI"] = "sfi";
    UnderlyingAsset["SGA"] = "sga";
    UnderlyingAsset["SGDX"] = "sgdx";
    UnderlyingAsset["SGR"] = "sgr";
    UnderlyingAsset["SGT"] = "sgt";
    UnderlyingAsset["SHDW"] = "shdw";
    UnderlyingAsset["SHEESH"] = "sheesh";
    UnderlyingAsset["SHIDO"] = "shido";
    UnderlyingAsset["SHK"] = "shk";
    UnderlyingAsset["SHOPX"] = "shopx";
    UnderlyingAsset["SHOW"] = "show";
    UnderlyingAsset["SHIB"] = "shib";
    UnderlyingAsset["SHR"] = "shr";
    UnderlyingAsset["SIH"] = "sih";
    UnderlyingAsset["SILV"] = "silv";
    UnderlyingAsset["SIPHER"] = "sipher";
    UnderlyingAsset["SIS"] = "sis";
    UnderlyingAsset["SKALE"] = "skale";
    UnderlyingAsset["SLAB"] = "slab";
    UnderlyingAsset["SLC"] = "slc";
    UnderlyingAsset["SLCL"] = "slcl";
    UnderlyingAsset["sol:slnd"] = "sol:slnd";
    UnderlyingAsset["SLOT"] = "slot";
    UnderlyingAsset["SLP"] = "slp";
    UnderlyingAsset["SLRS"] = "slrs";
    UnderlyingAsset["SLVX"] = "slvx";
    UnderlyingAsset["SMT"] = "smt";
    UnderlyingAsset["SNC"] = "snc";
    UnderlyingAsset["SNM"] = "snm";
    UnderlyingAsset["SNOV"] = "snov";
    UnderlyingAsset["SNT"] = "snt";
    UnderlyingAsset["SNX"] = "snx";
    UnderlyingAsset["SNY"] = "sny";
    UnderlyingAsset["SOC"] = "soc";
    UnderlyingAsset["SOHM"] = "sohm";
    UnderlyingAsset["SOMM"] = "somm";
    UnderlyingAsset["SOS"] = "sos";
    UnderlyingAsset["SPA"] = "spa";
    UnderlyingAsset["SPELL"] = "spell";
    UnderlyingAsset["SPF"] = "spf";
    UnderlyingAsset["SPO"] = "spo";
    UnderlyingAsset["SOLVE"] = "solve";
    UnderlyingAsset["SQUID2.0"] = "squid2.0";
    UnderlyingAsset["SRNT"] = "srnt";
    UnderlyingAsset["SRM"] = "srm";
    UnderlyingAsset["sol:srm"] = "sol:srm";
    UnderlyingAsset["STARS"] = "stars";
    UnderlyingAsset["STATE"] = "state";
    UnderlyingAsset["STBU"] = "stbu";
    UnderlyingAsset["STC"] = "stc";
    UnderlyingAsset["STCV2"] = "stcv2";
    UnderlyingAsset["STEP"] = "step";
    UnderlyingAsset["STETH"] = "steth";
    UnderlyingAsset["STG"] = "stg";
    UnderlyingAsset["STKAAVE"] = "stkaave";
    UnderlyingAsset["STMX"] = "stmx";
    UnderlyingAsset["STORE"] = "store";
    UnderlyingAsset["STORJ"] = "storj";
    UnderlyingAsset["STORM"] = "storm";
    UnderlyingAsset["STRIKE"] = "strike";
    UnderlyingAsset["STRK"] = "strk";
    UnderlyingAsset["STRONG"] = "strong";
    UnderlyingAsset["STSOL"] = "stsol";
    UnderlyingAsset["STZEN"] = "stzen";
    UnderlyingAsset["SUI-ERC20"] = "sui-erc20";
    UnderlyingAsset["SUN"] = "sun";
    UnderlyingAsset["SUNNY"] = "sunny";
    UnderlyingAsset["SUPER"] = "super";
    UnderlyingAsset["SUPERPERIO"] = "superperio";
    UnderlyingAsset["SUSD"] = "susd";
    UnderlyingAsset["SUSDE"] = "susde";
    UnderlyingAsset["SUSHI"] = "sushi";
    UnderlyingAsset["SQUIG"] = "squig";
    UnderlyingAsset["SVT"] = "svt";
    UnderlyingAsset["SWAG"] = "swag";
    UnderlyingAsset["SWEAT"] = "sweat";
    UnderlyingAsset["SWETH"] = "sweth";
    UnderlyingAsset["SWISE"] = "swice";
    UnderlyingAsset["SWITCH"] = "switch";
    UnderlyingAsset["SWRV"] = "swrv";
    UnderlyingAsset["SXP"] = "sxp";
    UnderlyingAsset["SYN"] = "syn";
    UnderlyingAsset["SYNCH"] = "synch";
    UnderlyingAsset["SYNTH-SUSD"] = "synth-susd";
    UnderlyingAsset["THRESHOLD"] = "threshold";
    UnderlyingAsset["THEU"] = "theu";
    UnderlyingAsset["TAUD"] = "taud";
    UnderlyingAsset["TBILL"] = "tbill";
    UnderlyingAsset["TBTC1"] = "tbtc1";
    UnderlyingAsset["TBTC2"] = "tbtc2";
    UnderlyingAsset["TCAD"] = "tcad";
    UnderlyingAsset["TCO"] = "tco";
    UnderlyingAsset["TEL"] = "tel";
    UnderlyingAsset["TELEGRAMDAO"] = "telegramdao";
    UnderlyingAsset["TEN"] = "ten";
    UnderlyingAsset["TENX"] = "tenx";
    UnderlyingAsset["TERC"] = "terc";
    UnderlyingAsset["TEUROC"] = "teuroc";
    UnderlyingAsset["TERC2DP"] = "terc2dp";
    UnderlyingAsset["TERC6DP"] = "terc6dp";
    UnderlyingAsset["TERC18DP"] = "terc18DP";
    UnderlyingAsset["TERC20"] = "terc20";
    UnderlyingAsset["TERC2DP1"] = "terc2dp1";
    UnderlyingAsset["TERC2DP2"] = "terc2dp2";
    UnderlyingAsset["TERC2DP3"] = "terc2dp3";
    UnderlyingAsset["TERC2DP4"] = "terc2dp4";
    UnderlyingAsset["TERC2DP5"] = "terc2dp5";
    UnderlyingAsset["TERC6DP1"] = "terc6dp1";
    UnderlyingAsset["TERC6DP2"] = "terc6dp2";
    UnderlyingAsset["TERC6DP3"] = "terc6dp3";
    UnderlyingAsset["TERC6DP4"] = "terc6dp4";
    UnderlyingAsset["TERC6DP5"] = "terc6dp5";
    UnderlyingAsset["TERC18DP1"] = "terc18dp1";
    UnderlyingAsset["TERC18DP2"] = "terc18dp2";
    UnderlyingAsset["TERC18DP3"] = "terc18dp3";
    UnderlyingAsset["TERC18DP4"] = "terc18dp4";
    UnderlyingAsset["TERC18DP5"] = "terc18dp5";
    UnderlyingAsset["TERC18DP6"] = "terc18dp6";
    UnderlyingAsset["TERC18DP7"] = "terc18dp7";
    UnderlyingAsset["TERC18DP8"] = "terc18dp8";
    UnderlyingAsset["TERC18DP9"] = "terc18dp9";
    UnderlyingAsset["TERC18DP10"] = "terc18dp10";
    UnderlyingAsset["TERC18DP11"] = "terc18dp11";
    UnderlyingAsset["TERC18DP12"] = "terc18dp12";
    UnderlyingAsset["TERC18DP13"] = "terc18dp13";
    UnderlyingAsset["TERC18DP14"] = "terc18dp14";
    UnderlyingAsset["TERC18DP15"] = "terc18dp15";
    UnderlyingAsset["BGERCH"] = "bgerch";
    UnderlyingAsset["TERM"] = "term";
    UnderlyingAsset["TGBP"] = "tgbp";
    UnderlyingAsset["TGOUSD"] = "tgousd";
    UnderlyingAsset["THKD"] = "thkd";
    UnderlyingAsset["TIO"] = "tio";
    UnderlyingAsset["TIOX"] = "tiox";
    UnderlyingAsset["TKMK"] = "tkmk";
    UnderlyingAsset["TKNT"] = "tknt";
    UnderlyingAsset["TKO"] = "tko";
    UnderlyingAsset["TKX"] = "tkx";
    UnderlyingAsset["TLAB"] = "tlab";
    UnderlyingAsset["TLM"] = "tlm";
    UnderlyingAsset["TLOS"] = "tlos";
    UnderlyingAsset["TMATIC"] = "tmatic";
    UnderlyingAsset["TMSN"] = "tmsn";
    UnderlyingAsset["TNT"] = "tnt";
    UnderlyingAsset["TOKAMAK"] = "tokamak";
    UnderlyingAsset["TOKE"] = "toke";
    UnderlyingAsset["TOMI"] = "tomi";
    UnderlyingAsset["TOMOBEAR"] = "tomobear";
    UnderlyingAsset["TOMOBEAR2"] = "tomobear2";
    UnderlyingAsset["TOMOBULL"] = "tomobull";
    UnderlyingAsset["TOMOE"] = "tomoe";
    UnderlyingAsset["TOK"] = "tok";
    UnderlyingAsset["TONCOIN"] = "toncoin";
    UnderlyingAsset["TOPM"] = "topm";
    UnderlyingAsset["TRAC"] = "trac";
    UnderlyingAsset["TRAXX"] = "traxx";
    UnderlyingAsset["TRIBE"] = "tribe";
    UnderlyingAsset["TRL"] = "trl";
    UnderlyingAsset["TRST"] = "trst";
    UnderlyingAsset["TRU"] = "tru";
    UnderlyingAsset["TRUF"] = "truf";
    UnderlyingAsset["TRUMPLOSE"] = "trumplose";
    UnderlyingAsset["TRUMPWIN"] = "trumpwin";
    UnderlyingAsset["TRXBEAR"] = "trxbear";
    UnderlyingAsset["TRXBULL"] = "trxbull";
    UnderlyingAsset["TRXDOOM"] = "trxdoom";
    UnderlyingAsset["TRX-ERC20"] = "TRX-ERC20";
    UnderlyingAsset["TRXHEDGE"] = "trxhedge";
    UnderlyingAsset["TRXMOON"] = "trxmoon";
    // Bilira
    UnderlyingAsset["TRYB"] = "tryb";
    // TRYB on Solana - https://solscan.io/token/6ry4WBDvAwAnrYJVv6MCog4J8zx6S3cPgSqnTsDZ73AR
    UnderlyingAsset["TRYB2"] = "tryb2";
    UnderlyingAsset["TRYBBEAR"] = "trybbear";
    UnderlyingAsset["TRYBBULL"] = "trybbull";
    UnderlyingAsset["TRYX"] = "tryx";
    UnderlyingAsset["TST"] = "tst";
    UnderlyingAsset["TSUKA"] = "tsuka";
    UnderlyingAsset["TULIP"] = "tulip";
    UnderlyingAsset["TUPOLIS"] = "tupolis";
    UnderlyingAsset["TUSD"] = "tusd";
    UnderlyingAsset["TUSDC"] = "tusdc";
    UnderlyingAsset["TUSDT"] = "tusdt";
    UnderlyingAsset["TUSRM"] = "tusrm";
    UnderlyingAsset["TWDOGE"] = "twdoge";
    UnderlyingAsset["TWETH"] = "tweth";
    UnderlyingAsset["TXL"] = "txl";
    UnderlyingAsset["TXSGD"] = "txsgd";
    UnderlyingAsset["TXUSD"] = "txusd";
    UnderlyingAsset["UAIR"] = "uair";
    UnderlyingAsset["UBXT"] = "ubxt";
    UnderlyingAsset["UCO"] = "uco";
    UnderlyingAsset["UKG"] = "ukg";
    UnderlyingAsset["UMA"] = "uma";
    UnderlyingAsset["UMEE"] = "umee";
    UnderlyingAsset["UNB"] = "unb";
    UnderlyingAsset["UNI"] = "uni";
    UnderlyingAsset["UOS"] = "uos";
    UnderlyingAsset["UP"] = "up";
    UnderlyingAsset["UPBTC"] = "upbtc";
    UnderlyingAsset["UPP"] = "upp";
    UnderlyingAsset["UPT"] = "upt";
    UnderlyingAsset["UPUSD"] = "upusd";
    UnderlyingAsset["UQC"] = "uqc";
    UnderlyingAsset["URHD"] = "urhd";
    UnderlyingAsset["sol:usdt"] = "sol:usdt";
    UnderlyingAsset["sol:usdc"] = "sol:usdc";
    UnderlyingAsset["USCC"] = "uscc";
    UnderlyingAsset["USDC"] = "usdc";
    UnderlyingAsset["USDC-POS-WORMHOLE"] = "usdc-pos-wormhole";
    UnderlyingAsset["USDE"] = "usde";
    UnderlyingAsset["USDGLO"] = "usdglo";
    UnderlyingAsset["USDH"] = "usdh";
    UnderlyingAsset["USDK"] = "usdk";
    // Also available on EOS
    UnderlyingAsset["USDT"] = "usdt";
    UnderlyingAsset["USDTBEAR"] = "usdtbear";
    UnderlyingAsset["USDTBULL"] = "usdtbull";
    UnderlyingAsset["USDTDOOM"] = "usdtdoom";
    UnderlyingAsset["USDTHEDGE"] = "usdthedge";
    UnderlyingAsset["USDTMOON"] = "usdtmoon";
    UnderlyingAsset["USDX"] = "usdx";
    UnderlyingAsset["USDY"] = "usdy";
    UnderlyingAsset["USG"] = "usg";
    UnderlyingAsset["USPX"] = "uspx";
    UnderlyingAsset["UST"] = "ust";
    UnderlyingAsset["USTB"] = "ustb";
    UnderlyingAsset["UST-WORMHOLE"] = "ust-wormhole";
    UnderlyingAsset["USX"] = "usx";
    UnderlyingAsset["USYC"] = "usyc";
    UnderlyingAsset["UTK"] = "utk";
    UnderlyingAsset["UTK1"] = "utk1";
    UnderlyingAsset["UXB"] = "uxb";
    UnderlyingAsset["UXP"] = "uxp";
    UnderlyingAsset["VALOR"] = "valor";
    UnderlyingAsset["VBNT"] = "vbnt";
    UnderlyingAsset["VCORE"] = "vcore";
    UnderlyingAsset["VDX"] = "vdx";
    UnderlyingAsset["VEC"] = "vec";
    UnderlyingAsset["VEE"] = "vee";
    UnderlyingAsset["VEGA"] = "vega";
    UnderlyingAsset["VEXT"] = "vext";
    UnderlyingAsset["VGX"] = "vgx";
    UnderlyingAsset["VI"] = "vi";
    UnderlyingAsset["VIB"] = "vib";
    UnderlyingAsset["VISR"] = "visr";
    UnderlyingAsset["VIU"] = "viu";
    UnderlyingAsset["VRA"] = "vra";
    UnderlyingAsset["VRGX"] = "vrgx";
    UnderlyingAsset["VSP"] = "vsp";
    UnderlyingAsset["VXC"] = "vxc";
    UnderlyingAsset["VXV"] = "vxv";
    UnderlyingAsset["W"] = "w";
    // Wrapped AAVE
    UnderlyingAsset["WAAVE"] = "waave";
    UnderlyingAsset["WABI"] = "wabi";
    UnderlyingAsset["WAFL"] = "wafl";
    UnderlyingAsset["WAGMI"] = "wagmi";
    // Wrapped AAVAX
    UnderlyingAsset["WAVAX"] = "wavax";
    UnderlyingAsset["WAVES"] = "waves";
    UnderlyingAsset["WAX"] = "wax";
    UnderlyingAsset["WAXP"] = "waxp";
    // Wrapped BNB
    UnderlyingAsset["WBNB"] = "wbnb";
    UnderlyingAsset["WECAN"] = "wecan";
    UnderlyingAsset["WFEE"] = "wfee";
    UnderlyingAsset["WHAT"] = "what";
    UnderlyingAsset["WOO"] = "woo";
    UnderlyingAsset["WTK"] = "wtk";
    UnderlyingAsset["WBTC"] = "wbtc";
    UnderlyingAsset["WDAIV2"] = "wdaiv2";
    UnderlyingAsset["WDOGE"] = "wdoge";
    UnderlyingAsset["WCFG"] = "wcfg";
    UnderlyingAsset["WEC"] = "wec";
    UnderlyingAsset["sol:wec"] = "sol:wec";
    UnderlyingAsset["WET"] = "wet";
    UnderlyingAsset["WETH"] = "weth";
    UnderlyingAsset["WEETH"] = "weeth";
    UnderlyingAsset["WFLOW"] = "wflow";
    UnderlyingAsset["WFFT"] = "wfft";
    UnderlyingAsset["WHALE"] = "whale";
    UnderlyingAsset["WHT"] = "wht";
    UnderlyingAsset["WILD"] = "wild";
    UnderlyingAsset["WNXM"] = "wnxm";
    UnderlyingAsset["WLD"] = "wld";
    UnderlyingAsset["WLUNA"] = "wluna";
    UnderlyingAsset["WLXT"] = "wlxt";
    // Wrapped SOL
    UnderlyingAsset["sol:wsol"] = "sol:wsol";
    UnderlyingAsset["WSTETH"] = "wsteth";
    UnderlyingAsset["WPX"] = "wpx";
    UnderlyingAsset["WTC"] = "wtc";
    // USD Coin (Wormhole)
    UnderlyingAsset["WUSDC"] = "wusdc";
    UnderlyingAsset["WUSDCV2"] = "wusdvcv2";
    // Tether USD (Wormhole)
    UnderlyingAsset["WUSDTV2"] = "wusdtv2";
    UnderlyingAsset["WXRP"] = "wxrp";
    UnderlyingAsset["WXRPV0"] = "wxrpv0";
    UnderlyingAsset["WXT"] = "wxt";
    UnderlyingAsset["XAUD"] = "xaud";
    UnderlyingAsset["XAURY"] = "xaury";
    UnderlyingAsset["XAUT"] = "xaut";
    UnderlyingAsset["XAUTBEAR"] = "xautbear";
    UnderlyingAsset["XAUTBULL"] = "xautbull";
    UnderlyingAsset["XBGOLD"] = "xbgold";
    UnderlyingAsset["XCD"] = "xcd";
    UnderlyingAsset["XCHNG"] = "xchng";
    UnderlyingAsset["XCN"] = "xcn";
    UnderlyingAsset["XDEFI"] = "xdefi";
    UnderlyingAsset["XDOGE"] = "xdoge";
    UnderlyingAsset["XEX"] = "xex";
    UnderlyingAsset["XLMBEAR"] = "xlmbear";
    UnderlyingAsset["XLMBULL"] = "xlmbull";
    UnderlyingAsset["XRL"] = "xrl";
    UnderlyingAsset["XRPBEAR"] = "xrpbear";
    UnderlyingAsset["XRPBULL"] = "xrpbull";
    UnderlyingAsset["XRPDOOM"] = "xrpdoom";
    UnderlyingAsset["XRPHEDGE"] = "xrphedge";
    UnderlyingAsset["XRPMOON"] = "xrpmoon";
    UnderlyingAsset["XSGD"] = "xsgd";
    UnderlyingAsset["XSUSHI"] = "xsushi";
    UnderlyingAsset["XTP"] = "xtp";
    UnderlyingAsset["XTZBEAR"] = "xtzbear";
    UnderlyingAsset["XTZBULL"] = "xtzbull";
    UnderlyingAsset["XUSD"] = "xusd";
    UnderlyingAsset["XX"] = "xx";
    UnderlyingAsset["XZK"] = "xzk";
    UnderlyingAsset["YAMV2"] = "yamv2";
    UnderlyingAsset["YFDAI"] = "yfdai";
    UnderlyingAsset["YFI"] = "yfi";
    UnderlyingAsset["YFII"] = "yfii";
    UnderlyingAsset["YFL"] = "yfl";
    UnderlyingAsset["YGG"] = "ygg";
    UnderlyingAsset["YLD"] = "yld";
    UnderlyingAsset["YNG"] = "yng";
    UnderlyingAsset["YSEY"] = "ysey";
    UnderlyingAsset["ZARX"] = "zarx";
    UnderlyingAsset["ZBC"] = "zbc";
    UnderlyingAsset["ZBU"] = "zbu";
    UnderlyingAsset["ZCO"] = "zco";
    UnderlyingAsset["ZECBEAR"] = "zecbear";
    UnderlyingAsset["ZECBULL"] = "zecbull";
    UnderlyingAsset["ZETAEVM"] = "zetaevm";
    UnderlyingAsset["ZIL"] = "zil";
    UnderlyingAsset["ZIP"] = "zip";
    UnderlyingAsset["ZIX"] = "zix";
    UnderlyingAsset["ZKL"] = "zkl";
    UnderlyingAsset["ZKS"] = "zks";
    UnderlyingAsset["ZLW"] = "zlw";
    UnderlyingAsset["ZMT"] = "zmt";
    UnderlyingAsset["ZOOM"] = "zoom";
    UnderlyingAsset["ZRO"] = "zro";
    UnderlyingAsset["ZRO-0x320"] = "zro-0x320";
    UnderlyingAsset["ZRO-0xFCF"] = "zro-0xfcf";
    UnderlyingAsset["ZRO-0xE5C"] = "zro-0xe5c";
    UnderlyingAsset["ZRX"] = "zrx";
    UnderlyingAsset["ZUSD"] = "zusd";
    UnderlyingAsset["xlm:BST-GADDFE4R72YUP2AOEL67OHZN3GJQYPC3VE734N2XFMEGRR2L32CZ3XYZ"] = "xlm:BST-GADDFE4R72YUP2AOEL67OHZN3GJQYPC3VE734N2XFMEGRR2L32CZ3XYZ";
    UnderlyingAsset["xlm:VELO-GDM4RQUQQUVSKQA7S6EM7XBZP3FCGH4Q7CL6TABQ7B2BEJ5ERARM2M5M"] = "xlm:VELO-GDM4RQUQQUVSKQA7S6EM7XBZP3FCGH4Q7CL6TABQ7B2BEJ5ERARM2M5M";
    UnderlyingAsset["xlm:SLT-GCKA6K5PCQ6PNF5RQBF7PQDJWRHO6UOGFMRLK3DYHDOI244V47XKQ4GP"] = "xlm:SLT-GCKA6K5PCQ6PNF5RQBF7PQDJWRHO6UOGFMRLK3DYHDOI244V47XKQ4GP";
    UnderlyingAsset["xlm:USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX"] = "xlm:USD-GDUKMGUGDZQK6YHYA5Z6AY2G4XDSZPSZ3SW5UN3ARVMO6QSRDWP5YLEX";
    UnderlyingAsset["xlm:ETH-GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5"] = "xlm:ETH-GBVOL67TMUQBGL4TZYNMY3ZQ5WGQYFPFD5VJRWXR72VA33VFNL225PL5";
    UnderlyingAsset["xlm:WXT-GASBLVHS5FOABSDNW5SPPH3QRJYXY5JHA2AOA2QHH2FJLZBRXSG4SWXT"] = "xlm:WXT-GASBLVHS5FOABSDNW5SPPH3QRJYXY5JHA2AOA2QHH2FJLZBRXSG4SWXT";
    UnderlyingAsset["xlm:USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN"] = "xlm:USDC-GA5ZSEJYB37JRC5AVCIA5MOP4RHTM335X2KGX3IHOJAPP5RE34K4KZVN";
    UnderlyingAsset["xlm:SIX-GDMS6EECOH6MBMCP3FYRYEVRBIV3TQGLOFQIPVAITBRJUMTI6V7A2X6Z"] = "xlm:SIX-GDMS6EECOH6MBMCP3FYRYEVRBIV3TQGLOFQIPVAITBRJUMTI6V7A2X6Z";
    UnderlyingAsset["xlm:BRLT-GCHQ3F2BF5P74DMDNOOGHT5DUCKC773AW5DTOFINC26W4KGYFPYDPRSO"] = "xlm:BRLT-GCHQ3F2BF5P74DMDNOOGHT5DUCKC773AW5DTOFINC26W4KGYFPYDPRSO";
    UnderlyingAsset["xlm:ARST-GCSAZVWXZKWS4XS223M5F54H2B6XPIIXZZGP7KEAIU6YSL5HDRGCI3DG"] = "xlm:ARST-GCSAZVWXZKWS4XS223M5F54H2B6XPIIXZZGP7KEAIU6YSL5HDRGCI3DG";
    UnderlyingAsset["xlm:AQUA-GBNZILSTVQZ4R7IKQDGHYGY2QXL5QOFJYQMXPKWRRM5PAV7Y4M67AQUA"] = "xlm:AQUA-GBNZILSTVQZ4R7IKQDGHYGY2QXL5QOFJYQMXPKWRRM5PAV7Y4M67AQUA";
    UnderlyingAsset["xlm:EURC-GDHU6WRG4IEQXM5NZ4BMPKOXHW76MZM4Y2IEMFDVXBSDP6SJY4ITNPP2"] = "xlm:EURC-GDHU6WRG4IEQXM5NZ4BMPKOXHW76MZM4Y2IEMFDVXBSDP6SJY4ITNPP2";
    UnderlyingAsset["xlm:GYEN-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB"] = "xlm:GYEN-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB";
    UnderlyingAsset["xlm:ZUSD-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB"] = "xlm:ZUSD-GDF6VOEGRWLOZ64PQQGKD2IYWA22RLT37GJKS2EJXZHT2VLAGWLC5TOB";
    UnderlyingAsset["xlm:EURS-GC5FGCDEOGOGSNWCCNKS3OMEVDHTE3Q5A5FEQWQKV3AXA7N6KDQ2CUZJ"] = "xlm:EURS-GC5FGCDEOGOGSNWCCNKS3OMEVDHTE3Q5A5FEQWQKV3AXA7N6KDQ2CUZJ";
    UnderlyingAsset["xlm:VEUR-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN"] = "xlm:VEUR-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN";
    UnderlyingAsset["xlm:VCHF-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN"] = "xlm:VCHF-GDXLSLCOPPHTWOQXLLKSVN4VN3G67WD2ENU7UMVAROEYVJLSPSEWXIZN";
    UnderlyingAsset["xlm:AUDD-GDC7X2MXTYSAKUUGAIQ7J7RPEIM7GXSAIWFYWWH4GLNFECQVJJLB2EEU"] = "xlm:AUDD-GDC7X2MXTYSAKUUGAIQ7J7RPEIM7GXSAIWFYWWH4GLNFECQVJJLB2EEU";
    // Eth NFTs
    // generic NFTs
    UnderlyingAsset["erc721:token"] = "erc721:token";
    UnderlyingAsset["erc1155:token"] = "erc1155:token";
    UnderlyingAsset["nonstandard:token"] = "nonstandard:token";
    // Test Eth NFTs
    UnderlyingAsset["terc721:token"] = "terc721:token";
    UnderlyingAsset["terc1155:token"] = "terc1155:token";
    UnderlyingAsset["tnonstandard:token"] = "tnonstandard:token";
    // Algorand mainnet tokens
    UnderlyingAsset["algo:USDC-31566704"] = "algo:USDC-31566704";
    UnderlyingAsset["algo:USDt-312769"] = "algo:USDt-312769";
    UnderlyingAsset["algo:MCAU-6547014"] = "algo:MCAU-6547014";
    UnderlyingAsset["algo:QCAD-84507107"] = "algo:QCAD-84507107";
    UnderlyingAsset["algo:VCAD-438505559"] = "algo:VCAD-438505559";
    // Kovan-only ERC20 tokens
    UnderlyingAsset["TEST"] = "test";
    UnderlyingAsset["SCHZ"] = "schz";
    UnderlyingAsset["CAT"] = "cat";
    // Stellar testnet tokens
    UnderlyingAsset["txlm:BST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L"] = "txlm:BST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L";
    UnderlyingAsset["txlm:TST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L"] = "txlm:TST-GBQTIOS3XGHB7LVYGBKQVJGCZ3R4JL5E4CBSWJ5ALIJUHBKS6263644L";
    // Algorand testnet tokens
    UnderlyingAsset["talgo:USON-16026728"] = "talgo:USON-16026728";
    UnderlyingAsset["talgo:SPRW-16026732"] = "talgo:SPRW-16026732";
    UnderlyingAsset["talgo:KAL-16026733"] = "talgo:KAL-16026733";
    UnderlyingAsset["talgo:USDC-10458941"] = "talgo:USDC-10458941";
    UnderlyingAsset["talgo:USDt-180447"] = "talgo:USDt-180447";
    UnderlyingAsset["talgo:JPT-162085446"] = "talgo:JPT-162085446";
    // EOS tokens
    UnderlyingAsset["CHEX"] = "chex";
    UnderlyingAsset["IQ"] = "iq";
    UnderlyingAsset["EOS_BOX"] = "eos:box";
    // Avax Token ERC-20
    UnderlyingAsset["avaxc:qi"] = "avaxc:qi";
    UnderlyingAsset["avaxc:xava"] = "avaxc:xava";
    UnderlyingAsset["avaxc:klo"] = "avaxc:klo";
    UnderlyingAsset["avaxc:joe"] = "avaxc:joe";
    UnderlyingAsset["avaxc:png"] = "avaxc:png";
    UnderlyingAsset["avaxc:usdt"] = "avaxc:usdt";
    UnderlyingAsset["avaxc:usdc"] = "avaxc:usdc";
    UnderlyingAsset["avaxc:link"] = "avaxc:link";
    UnderlyingAsset["avaxc:cai"] = "avaxc:cai";
    UnderlyingAsset["avaxc:aave"] = "avaxc:aave";
    UnderlyingAsset["avaxc:btc"] = "avaxc:btc";
    UnderlyingAsset["avaxc:dai"] = "avaxc:dai";
    UnderlyingAsset["avaxc:tryb"] = "avaxc:tryb";
    UnderlyingAsset["avaxc:wbtc"] = "avaxc:wbtc";
    UnderlyingAsset["avaxc:weth"] = "avaxc:weth";
    UnderlyingAsset["avaxc:sbc"] = "avaxc:sbc";
    UnderlyingAsset["avaxc:xsgd"] = "avaxc:xsgd";
    UnderlyingAsset["tavaxc:opm"] = "tavaxc:opm";
    UnderlyingAsset["tavaxc:cop2peq"] = "tavaxc:cop2peq";
    UnderlyingAsset["tavaxc:xsgd"] = "tavaxc:xsgd";
    // Begin FTX missing AVAXC tokens
    UnderlyingAsset["avaxc:yeti"] = "avaxc:yeti";
    UnderlyingAsset["avaxc:spell"] = "avaxc:spell";
    UnderlyingAsset["avaxc:yusd"] = "avaxc:yusd";
    UnderlyingAsset["avaxc:yusdcrv-f"] = "avaxc:yusdcrv-f";
    UnderlyingAsset["avaxc:ecd"] = "avaxc:ecd";
    UnderlyingAsset["avaxc:blzz"] = "avaxc:blzz";
    UnderlyingAsset["avaxc:ptp"] = "avaxc:ptp";
    UnderlyingAsset["avaxc:stg"] = "avaxc:stg";
    UnderlyingAsset["avaxc:syn"] = "avaxc:syn";
    UnderlyingAsset["avaxc:aavausdc"] = "avaxc:aavausdc";
    UnderlyingAsset["avaxc:tusd"] = "avaxc:tusd";
    UnderlyingAsset["avaxc:crv"] = "avaxc:crv";
    UnderlyingAsset["avaxc:savax"] = "avaxc:savax";
    UnderlyingAsset["avaxc:ampl"] = "avaxc:ampl";
    UnderlyingAsset["avaxc:cnr"] = "avaxc:cnr";
    UnderlyingAsset["avaxc:roco"] = "avaxc:roco";
    UnderlyingAsset["avaxc:aavadai"] = "avaxc:aavadai";
    UnderlyingAsset["avaxc:vtx"] = "avaxc:vtx";
    UnderlyingAsset["avaxc:wavax"] = "avaxc:wavax";
    UnderlyingAsset["avaxc:bnb"] = "avaxc:bnb";
    UnderlyingAsset["avaxc:aavausdt"] = "avaxc:aavausdt";
    UnderlyingAsset["avaxc:acre"] = "avaxc:acre";
    UnderlyingAsset["avaxc:gmx"] = "avaxc:gmx";
    UnderlyingAsset["avaxc:mim"] = "avaxc:mim";
    UnderlyingAsset["avaxc:axlusdc"] = "avaxc:axlusdc";
    UnderlyingAsset["avaxc:lot"] = "avaxc:lot";
    UnderlyingAsset["avaxc:av3crv"] = "avaxc:av3crv";
    UnderlyingAsset["avaxc:time"] = "avaxc:time";
    UnderlyingAsset["avaxc:uni.e"] = "avaxc:uni.e";
    UnderlyingAsset["avaxc:sb"] = "avaxc:sb";
    UnderlyingAsset["avaxc:dyp"] = "avaxc:dyp";
    UnderlyingAsset["avaxc:sing"] = "avaxc:sing";
    UnderlyingAsset["avaxc:gohm"] = "avaxc:gohm";
    UnderlyingAsset["avaxc:boofi"] = "avaxc:boofi";
    UnderlyingAsset["avaxc:eth"] = "avaxc:eth";
    UnderlyingAsset["avaxc:wmemo"] = "avaxc:wmemo";
    UnderlyingAsset["avaxc:fxs"] = "avaxc:fxs";
    UnderlyingAsset["avaxc:sifu"] = "avaxc:sifu";
    UnderlyingAsset["avaxc:sushi.e"] = "avaxc:sushi.e";
    UnderlyingAsset["avaxc:sushi"] = "avaxc:sushi";
    UnderlyingAsset["avaxc:mimatic"] = "avaxc:mimatic";
    UnderlyingAsset["avaxc:sspell"] = "avaxc:sspell";
    UnderlyingAsset["avaxc:grape"] = "avaxc:grape";
    UnderlyingAsset["avaxc:xjoe"] = "avaxc:xjoe";
    UnderlyingAsset["avaxc:bsgg"] = "avaxc:bsgg";
    UnderlyingAsset["avaxc:roy"] = "avaxc:roy";
    UnderlyingAsset["avaxc:wow"] = "avaxc:wow";
    UnderlyingAsset["avaxc:wine"] = "avaxc:wine";
    UnderlyingAsset["avaxc:mu"] = "avaxc:mu";
    UnderlyingAsset["avaxc:frax"] = "avaxc:frax";
    UnderlyingAsset["avaxc:movr"] = "avaxc:movr";
    UnderlyingAsset["avaxc:ice"] = "avaxc:ice";
    UnderlyingAsset["avaxc:note"] = "avaxc:note";
    // End FTX missing AVAXC tokens
    // polygon Token ERC-20
    UnderlyingAsset["polygon:usdc"] = "polygon:usdc";
    UnderlyingAsset["polygon:usdcv2"] = "polygon:usdcv2";
    UnderlyingAsset["polygon:usdt"] = "polygon:usdt";
    UnderlyingAsset["polygon:weth"] = "polygon:weth";
    UnderlyingAsset["polygon:wbtc"] = "polygon:wbtc";
    UnderlyingAsset["polygon:sand"] = "polygon:sand";
    UnderlyingAsset["polygon:dai"] = "polygon:dai";
    UnderlyingAsset["polygon:woo"] = "polygon:woo";
    UnderlyingAsset["polygon:aave"] = "polygon:aave";
    UnderlyingAsset["polygon:link"] = "polygon:link";
    UnderlyingAsset["polygon:tusd"] = "polygon:tusd";
    UnderlyingAsset["polygon:cel"] = "polygon:cel";
    UnderlyingAsset["polygon:busd"] = "polygon:busd";
    UnderlyingAsset["polygon:frax"] = "polygon:frax";
    UnderlyingAsset["polygon:crv"] = "polygon:crv";
    UnderlyingAsset["polygon:uni"] = "polygon:uni";
    UnderlyingAsset["polygon:fcd"] = "polygon:fcd";
    UnderlyingAsset["polygon:ape"] = "polygon:ape";
    UnderlyingAsset["polygon:srm"] = "polygon:srm";
    UnderlyingAsset["polygon:fly"] = "polygon:fly";
    UnderlyingAsset["polygon:gfc"] = "polygon:gfc";
    UnderlyingAsset["polygon:rbw"] = "polygon:rbw";
    UnderlyingAsset["polygon:zed"] = "polygon:zed";
    UnderlyingAsset["polygon:vext"] = "polygon:vext";
    UnderlyingAsset["polygon:vcnt"] = "polygon:vcnt";
    UnderlyingAsset["polygon:sushi"] = "polygon:sushi";
    UnderlyingAsset["polygon:wmatic"] = "polygon:wmatic";
    UnderlyingAsset["polygon:1inch"] = "polygon:1inch";
    UnderlyingAsset["polygon:comp"] = "polygon:comp";
    UnderlyingAsset["polygon:sol"] = "polygon:sol";
    UnderlyingAsset["polygon:wavax"] = "polygon:wavax";
    UnderlyingAsset["polygon:wbnb"] = "polygon:wbnb";
    UnderlyingAsset["polygon:wftm"] = "polygon:wftm";
    UnderlyingAsset["polygon:yfi"] = "polygon:yfi";
    UnderlyingAsset["polygon:treta"] = "polygon:treta";
    UnderlyingAsset["polygon:orb"] = "polygon:orb";
    UnderlyingAsset["polygon:route"] = "polygon:route";
    UnderlyingAsset["polygon:sbc"] = "polygon:sbc";
    UnderlyingAsset["polygon:xsgd"] = "polygon:xsgd";
    UnderlyingAsset["polygon:dimo"] = "polygon:dimo";
    UnderlyingAsset["polygon:bcut"] = "polygon:bcut";
    UnderlyingAsset["polygon:pme"] = "polygon:pme";
    UnderlyingAsset["polygon:dipe"] = "polygon:dipe";
    UnderlyingAsset["polygon:lif3"] = "polygon:lif3";
    UnderlyingAsset["polygon:l3usd"] = "polygon:l3usd";
    UnderlyingAsset["polygon:moca"] = "polygon:moca";
    UnderlyingAsset["polygon:mask"] = "polygon:mask";
    UnderlyingAsset["polygon:nexo"] = "polygon:nexo";
    UnderlyingAsset["polygon:om"] = "polygon:om";
    UnderlyingAsset["polygon:pyr"] = "polygon:pyr";
    UnderlyingAsset["polygon:renbtc"] = "polygon:renbtc";
    UnderlyingAsset["polygon:req"] = "polygon:req";
    UnderlyingAsset["polygon:rndr"] = "polygon:rndr";
    UnderlyingAsset["polygon:snx"] = "polygon:snx";
    UnderlyingAsset["polygon:trb"] = "polygon:trb";
    UnderlyingAsset["polygon:ali"] = "polygon:ali";
    UnderlyingAsset["polygon:bal"] = "polygon:bal";
    UnderlyingAsset["polygon:elon"] = "polygon:elon";
    UnderlyingAsset["polygon:hex"] = "polygon:hex";
    UnderlyingAsset["polygon:iotx"] = "polygon:iotx";
    UnderlyingAsset["polygon:agix"] = "polygon:agix";
    UnderlyingAsset["polygon:avax"] = "polygon:avax";
    UnderlyingAsset["polygon:band"] = "polygon:band";
    UnderlyingAsset["polygon:blz"] = "polygon:blz";
    UnderlyingAsset["polygon:bnb"] = "polygon:bnb";
    UnderlyingAsset["polygon:bnt"] = "polygon:bnt";
    UnderlyingAsset["polygon:chz"] = "polygon:chz";
    UnderlyingAsset["polygon:enj"] = "polygon:enj";
    UnderlyingAsset["polygon:fet"] = "polygon:fet";
    UnderlyingAsset["polygon:forth"] = "polygon:forth";
    UnderlyingAsset["polygon:glm"] = "polygon:glm";
    UnderlyingAsset["polygon:gno"] = "polygon:gno";
    UnderlyingAsset["polygon:gohm"] = "polygon:gohm";
    UnderlyingAsset["polygon:gtc"] = "polygon:gtc";
    UnderlyingAsset["polygon:gusd"] = "polygon:gusd";
    UnderlyingAsset["polygon:hot"] = "polygon:hot";
    UnderlyingAsset["polygon:inj"] = "polygon:inj";
    UnderlyingAsset["polygon:lit"] = "polygon:lit";
    UnderlyingAsset["polygon:lrc"] = "polygon:lrc";
    UnderlyingAsset["polygon:mana"] = "polygon:mana";
    UnderlyingAsset["polygon:shib"] = "polygon:shib";
    UnderlyingAsset["polygon:sxp"] = "polygon:sxp";
    // Polygon NFTs
    // generic NFTs
    UnderlyingAsset["erc721:polygontoken"] = "erc721:polygontoken";
    UnderlyingAsset["erc1155:polygontoken"] = "erc1155:polygontoken";
    // BSC Token BEP-20
    UnderlyingAsset["bsc:busd"] = "bsc:busd";
    UnderlyingAsset["bsc:hook"] = "bsc:hook";
    UnderlyingAsset["bsc:ksm"] = "bsc:ksm";
    UnderlyingAsset["bsc:usdt"] = "bsc:usdt";
    UnderlyingAsset["bsc:vet"] = "bsc:vet";
    UnderlyingAsset["bsc:cake"] = "bsc:cake";
    UnderlyingAsset["bsc:litt"] = "bsc:litt";
    UnderlyingAsset["bsc:xvs"] = "bsc:xvs";
    UnderlyingAsset["bsc:epx"] = "bsc:epx";
    UnderlyingAsset["bsc:usdc"] = "bsc:usdc";
    UnderlyingAsset["bsc:eth"] = "bsc:eth";
    UnderlyingAsset["bsc:dd"] = "bsc:dd";
    UnderlyingAsset["bsc:ltc"] = "bsc:ltc";
    UnderlyingAsset["bsc:mask"] = "bsc:mask";
    UnderlyingAsset["bsc:matic"] = "bsc:matic";
    UnderlyingAsset["bsc:mbox"] = "bsc:mbox";
    UnderlyingAsset["bsc:mdt"] = "bsc:mdt";
    UnderlyingAsset["bsc:nuls"] = "bsc:nuls";
    UnderlyingAsset["bsc:ont"] = "bsc:ont";
    UnderlyingAsset["bsc:orn"] = "bsc:orn";
    UnderlyingAsset["bsc:porto"] = "bsc:porto";
    UnderlyingAsset["bsc:reef"] = "bsc:reef";
    UnderlyingAsset["bsc:renbtc"] = "bsc:renbtc";
    UnderlyingAsset["bsc:snx"] = "bsc:snx";
    UnderlyingAsset["bsc:tking"] = "bsc:tking";
    UnderlyingAsset["bsc:tlm"] = "bsc:tlm";
    UnderlyingAsset["bsc:ton"] = "bsc:ton";
    UnderlyingAsset["bsc:trx"] = "bsc:trx";
    UnderlyingAsset["bsc:wbnb"] = "bsc:wbnb";
    UnderlyingAsset["bsc:win"] = "bsc:win";
    UnderlyingAsset["bsc:wrx"] = "bsc:wrx";
    UnderlyingAsset["bsc:yfii"] = "bsc:yfii";
    UnderlyingAsset["bsc:zil"] = "bsc:zil";
    UnderlyingAsset["bsc:1inch"] = "bsc:1inch";
    UnderlyingAsset["bsc:ada"] = "bsc:ada";
    UnderlyingAsset["bsc:alice"] = "bsc:alice";
    UnderlyingAsset["bsc:ankr"] = "bsc:ankr";
    UnderlyingAsset["bsc:avax"] = "bsc:avax";
    UnderlyingAsset["bsc:beta"] = "bsc:beta";
    UnderlyingAsset["bsc:btt"] = "bsc:btt";
    UnderlyingAsset["bsc:celr"] = "bsc:celr";
    UnderlyingAsset["bsc:chr"] = "bsc:chr";
    UnderlyingAsset["bsc:coti"] = "bsc:coti";
    UnderlyingAsset["bsc:cream"] = "bsc:cream";
    UnderlyingAsset["bsc:dar"] = "bsc:dar";
    UnderlyingAsset["bsc:degov2"] = "bsc:degov2";
    UnderlyingAsset["bsc:dodo"] = "bsc:dodo";
    UnderlyingAsset["bsc:elon"] = "bsc:elon";
    UnderlyingAsset["bsc:etc"] = "bsc:etc";
    UnderlyingAsset["bsc:front"] = "bsc:front";
    UnderlyingAsset["bsc:hft"] = "bsc:hft";
    UnderlyingAsset["bsc:high"] = "bsc:high";
    UnderlyingAsset["bsc:inj"] = "bsc:inj";
    UnderlyingAsset["bsc:iotx"] = "bsc:iotx";
    UnderlyingAsset["bsc:auto"] = "bsc:auto";
    UnderlyingAsset["bsc:fet"] = "bsc:fet";
    UnderlyingAsset["bsc:kas"] = "bsc:kas";
    UnderlyingAsset["bsc:lit"] = "bsc:lit";
    UnderlyingAsset["bsc:mana"] = "bsc:mana";
    UnderlyingAsset["bsc:shib"] = "bsc:shib";
    UnderlyingAsset["bsc:sxp"] = "bsc:sxp";
    UnderlyingAsset["bsc:nnn"] = "bsc:nnn";
    UnderlyingAsset["bsc:nvm"] = "bsc:nvm";
    UnderlyingAsset["tbsc:busd"] = "tbsc:busd";
    // BSC NFTs
    // generic NFTs
    UnderlyingAsset["erc721:bsctoken"] = "erc721:bsctoken";
    UnderlyingAsset["erc1155:bsctoken"] = "erc1155:bsctoken";
    // Test BSC NFTs
    UnderlyingAsset["terc721:bsctoken"] = "terc721:bsctoken";
    UnderlyingAsset["terc1155:bsctoken"] = "terc1155:bsctoken";
    // Polygon testnet tokens
    UnderlyingAsset["tpolygon:derc20"] = "tpolygon:derc20";
    UnderlyingAsset["tpolygon:link"] = "tpolygon:link";
    UnderlyingAsset["tpolygon:name"] = "tpolygon:name";
    UnderlyingAsset["tpolygon:opm"] = "tpolygon:opm";
    UnderlyingAsset["tpolygon:pme"] = "tpolygon:pme";
    UnderlyingAsset["tpolygon:xsgd"] = "tpolygon:xsgd";
    UnderlyingAsset["tpolygon:terc18dp"] = "tpolygon:terc18dp";
    UnderlyingAsset["tpolygon:terc6dp"] = "tpolygon:terc6dp";
    UnderlyingAsset["tpolygon:usdt"] = "tpolygon:usdt";
    UnderlyingAsset["tpolygon:usdc"] = "tpolygon:usdc";
    // generic NFTs
    UnderlyingAsset["terc721:polygontoken"] = "terc721:polygontoken";
    UnderlyingAsset["terc1155:polygontoken"] = "terc1155:polygontoken";
    // Arbitrum mainnet tokens
    UnderlyingAsset["arbeth:link"] = "arbeth:link";
    UnderlyingAsset["arbeth:usdc"] = "arbeth:usdc";
    UnderlyingAsset["arbeth:usdcv2"] = "arbeth:usdcv2";
    UnderlyingAsset["arbeth:usdt"] = "arbeth:usdt";
    UnderlyingAsset["arbeth:arb"] = "arbeth:arb";
    UnderlyingAsset["arbeth:sqd"] = "arbeth:sqd";
    UnderlyingAsset["arbeth:w"] = "arbeth:w";
    UnderlyingAsset["arbeth:comp"] = "arbeth:comp";
    UnderlyingAsset["arbeth:coti"] = "arbeth:coti";
    UnderlyingAsset["arbeth:gno"] = "arbeth:gno";
    UnderlyingAsset["arbeth:gohm"] = "arbeth:gohm";
    UnderlyingAsset["arbeth:grt"] = "arbeth:grt";
    UnderlyingAsset["arbeth:knc"] = "arbeth:knc";
    UnderlyingAsset["arbeth:trb"] = "arbeth:trb";
    UnderlyingAsset["arbeth:tusd"] = "arbeth:tusd";
    UnderlyingAsset["arbeth:uma"] = "arbeth:uma";
    UnderlyingAsset["arbeth:uni"] = "arbeth:uni";
    UnderlyingAsset["arbeth:weth"] = "arbeth:weth";
    UnderlyingAsset["arbeth:woo"] = "arbeth:woo";
    UnderlyingAsset["arbeth:yfi"] = "arbeth:yfi";
    // Arbitrum testnet tokens
    UnderlyingAsset["tarbeth:link"] = "tarbeth:link";
    // Optimism mainnet tokens
    UnderlyingAsset["opeth:link"] = "opeth:link";
    UnderlyingAsset["opeth:usdc"] = "opeth:usdc";
    UnderlyingAsset["opeth:usdcv2"] = "opeth:usdcv2";
    UnderlyingAsset["opeth:usdt"] = "opeth:usdt";
    UnderlyingAsset["opeth:op"] = "opeth:op";
    // Optimism testnet tokens
    UnderlyingAsset["topeth:terc18dp"] = "topeth:terc18dp";
    // zkSync mainnet tokens
    UnderlyingAsset["zketh:link"] = "zketh:link";
    // zkSync testnet tokens
    UnderlyingAsset["tzketh:link"] = "tzketh:link";
    UnderlyingAsset["ERC721"] = "erc721";
    UnderlyingAsset["ERC1155"] = "erc1155";
    UnderlyingAsset["NONSTANDARD"] = "nonstandard";
    // Cardano Token
    UnderlyingAsset["adaTestnetToken"] = "temporary-placeholder";
    // solana token
    UnderlyingAsset["3uejh-usdc"] = "3uejh-usdc";
    UnderlyingAsset["avax-usdc"] = "avax-usdc";
    UnderlyingAsset["bop-usdc"] = "bop-usdc";
    UnderlyingAsset["sol:crown"] = "sol:crown";
    UnderlyingAsset["elu-usdt"] = "elu-usdt";
    UnderlyingAsset["fida-usdc"] = "fida-usdc";
    UnderlyingAsset["fida-usdt"] = "fida-usdt";
    UnderlyingAsset["ftt-ftt"] = "ftt-ftt";
    UnderlyingAsset["link-usdc"] = "link-usdc";
    UnderlyingAsset["lqid-usdc"] = "lqid-usdc";
    UnderlyingAsset["maticpo-usdc"] = "maticpo-usdc";
    UnderlyingAsset["msol-sol"] = "msol-sol";
    UnderlyingAsset["msol-usdc"] = "msol-usdc";
    UnderlyingAsset["prism-usdc"] = "prism-usdc";
    UnderlyingAsset["sol:pyth"] = "sol:pyth";
    UnderlyingAsset["rendoge-usdc"] = "rendoge-usdc";
    UnderlyingAsset["shdw-usdc"] = "shdw-usdc";
    UnderlyingAsset["sol-wtust"] = "sol-wtust";
    UnderlyingAsset["srm-usdc"] = "srm-usdc";
    UnderlyingAsset["srmet-srm"] = "srmet-srm";
    UnderlyingAsset["sushi-usdc"] = "sushi-usdc";
    UnderlyingAsset["tuatlas"] = "tuatlas";
    UnderlyingAsset["tucope"] = "tucope";
    UnderlyingAsset["tulike"] = "tulike";
    UnderlyingAsset["tureal"] = "tureal";
    UnderlyingAsset["tusamo"] = "tusamo";
    UnderlyingAsset["usdt-usdc"] = "usdt-usdc";
    UnderlyingAsset["wbwbnb-usdc"] = "wbwbnb-usdc";
    UnderlyingAsset["wheth-usdc"] = "wheth-usdc";
    UnderlyingAsset["wtust-usdt"] = "wtust-usdt";
    UnderlyingAsset["xcope-usdc"] = "xcope-usdc";
    UnderlyingAsset["xrp-sollet"] = "xrp-sollet";
    UnderlyingAsset["aury"] = "aury";
    UnderlyingAsset["dio"] = "dio";
    UnderlyingAsset["sol-perp"] = "sol-perp";
    UnderlyingAsset["sol-woo"] = "sol-woo";
    UnderlyingAsset["sol-weth"] = "sol-weth";
    UnderlyingAsset["btc-sollet"] = "btc-sollet";
    UnderlyingAsset["eth-sollet"] = "eth-sollet";
    UnderlyingAsset["sol:bonk"] = "sol:bonk";
    UnderlyingAsset["jto"] = "jto";
    UnderlyingAsset["jup"] = "jup";
    UnderlyingAsset["sol:honey"] = "sol:honey";
    UnderlyingAsset["mobile"] = "mobile";
    UnderlyingAsset["wif"] = "wif";
    UnderlyingAsset["natix"] = "natix";
    UnderlyingAsset["sol:ks"] = "sol:ks";
    UnderlyingAsset["sol:apusdt"] = "sol:apusdt";
    UnderlyingAsset["sol:acusd"] = "sol:acusd";
    UnderlyingAsset["sol:solink"] = "sol:solink";
    UnderlyingAsset["sol:block"] = "sol:block";
    UnderlyingAsset["sol:render"] = "sol:render";
    UnderlyingAsset["sol:wen"] = "sol:wen";
    UnderlyingAsset["sol:mew"] = "sol:mew";
    UnderlyingAsset["sol:moveusd"] = "sol:moveusd";
    UnderlyingAsset["sol:pyusd"] = "sol:pyusd";
    // TRX tokens
    UnderlyingAsset["trx:jst"] = "trx:jst";
    UnderlyingAsset["trx:tusd"] = "trx:tusd";
    UnderlyingAsset["trx:win"] = "trx:win";
    // XRP tokens
    UnderlyingAsset["txrp:tst-rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd"] = "txrp:tst-rP9jPyP5kyvFRb6ZiRghAGw5u8SGAmU4bd";
    // Sui tokens
    UnderlyingAsset["sui:deep"] = "sui:deep";
    // Sui testnet tokens
    UnderlyingAsset["tsui:deep"] = "tsui:deep";
})(UnderlyingAsset = exports.UnderlyingAsset || (exports.UnderlyingAsset = {}));
/**
 * This is the curve BitGo signs against with the user, backup and BitGo key.
 */
var KeyCurve;
(function (KeyCurve) {
    KeyCurve["Secp256k1"] = "secp256k1";
    KeyCurve["Ed25519"] = "ed25519";
    KeyCurve["BLS"] = "bls";
})(KeyCurve = exports.KeyCurve || (exports.KeyCurve = {}));
/**
 * This enum contains the base units for the coins that BitGo supports
 */
var BaseUnit;
(function (BaseUnit) {
    BaseUnit["ATOM"] = "uatom";
    BaseUnit["ETH"] = "wei";
    BaseUnit["BTC"] = "satoshi";
    BaseUnit["BSC"] = "jager";
    BaseUnit["XLM"] = "stroop";
    BaseUnit["TRX"] = "sun";
    BaseUnit["HBAR"] = "tinybar";
    BaseUnit["ALGO"] = "microAlgo";
    BaseUnit["EOS"] = "eos";
    BaseUnit["SOL"] = "lamport";
    BaseUnit["ADA"] = "lovelace";
    BaseUnit["USD"] = "USD";
    BaseUnit["LNBTC"] = "millisatoshi";
    BaseUnit["LTC"] = "microlitecoins";
    BaseUnit["DASH"] = "duff";
    BaseUnit["ZEC"] = "zatoshi";
    BaseUnit["CSPR"] = "mote";
    BaseUnit["DOT"] = "planck";
    BaseUnit["XRP"] = "drop";
    BaseUnit["XTZ"] = "micro xtz";
    BaseUnit["STX"] = "micro-STX";
    BaseUnit["SUI"] = "MIST";
    BaseUnit["TON"] = "nanoton";
    BaseUnit["NEAR"] = "yocto";
    BaseUnit["OFC"] = "ofcCoin";
    BaseUnit["OSMO"] = "uosmo";
    BaseUnit["FIAT"] = "fiatCoin";
    BaseUnit["TIA"] = "utia";
    BaseUnit["HASH"] = "nhash";
    BaseUnit["BLD"] = "ubld";
    BaseUnit["SEI"] = "usei";
    BaseUnit["INJECTIVE"] = "inj";
    BaseUnit["ZETA"] = "azeta";
    BaseUnit["KAVA"] = "ukava";
    BaseUnit["COREUM"] = "ucore";
    BaseUnit["TCOREUM"] = "utestcore";
    BaseUnit["BERA"] = "abera";
    BaseUnit["ISLM"] = "aISLM";
})(BaseUnit = exports.BaseUnit || (exports.BaseUnit = {}));
class BaseCoin {
    constructor(options) {
        this.validateOptions(options);
        this.id = options.id;
        this.fullName = options.fullName;
        this.name = options.name;
        this.alias = options.alias;
        this.prefix = options.prefix;
        this.suffix = options.suffix;
        this.baseUnit = options.baseUnit;
        this.kind = options.kind;
        this.family = options.network.family;
        this.isToken = options.isToken;
        this.features = options.features;
        this.decimalPlaces = options.decimalPlaces;
        this.asset = options.asset;
        this.network = options.network;
        this.primaryKeyCurve = options.primaryKeyCurve;
    }
    /**
     * Ensures that the base coin constructor was passed a valid set of options.
     *
     * This includes checking that:
     * - All coin features of the new instance are allowed by the coin class
     * - No features required by the coin class are missing from the new instance
     * @param {BaseCoinConstructorOptions} options
     * @throws {DisallowedCoinFeatureError} if any of the coin features are not allowed for the coin class
     * @throws {MissingRequiredCoinFeatureError} if any features required by the coin class are missing
     */
    validateOptions(options) {
        const requiredFeatures = this.requiredFeatures();
        const disallowedFeatures = this.disallowedFeatures();
        const intersectionFeatures = Array.from(requiredFeatures).filter((feat) => disallowedFeatures.has(feat));
        if (intersectionFeatures.length > 0) {
            throw new errors_1.ConflictingCoinFeaturesError(options.name, intersectionFeatures);
        }
        for (const feature of options.features) {
            if (disallowedFeatures.has(feature)) {
                throw new errors_1.DisallowedCoinFeatureError(options.name, feature);
            }
            if (requiredFeatures.has(feature)) {
                requiredFeatures.delete(feature);
            }
        }
        if (requiredFeatures.size > 0) {
            // some required features were missing
            throw new errors_1.MissingRequiredCoinFeatureError(options.name, Array.from(requiredFeatures));
        }
        // assets require a valid uuid v4 id
        if (!BaseCoin.isValidUuidV4(options.id)) {
            throw new errors_1.InvalidIdError(options.name, options.id);
        }
    }
}
exports.BaseCoin = BaseCoin;
BaseCoin.isValidUuidV4 = (uuid) => {
    const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
    return uuidRegex.test(uuid);
};
