// @bitgo/sdk-lib-mpc/dist/src/openssl/openssl.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpenSSL = void 0;
const wasi_1 = require("@wasmer/wasi");
class OpenSSL {
    constructor() {
        this.isInitialized = false;
    }
    async init(openSSLBytes) {
        await (0, wasi_1.init)();
        this.waModule = await WebAssembly.compile(openSSLBytes);
        this.isInitialized = true;
    }
    async generateSafePrime(bitLength) {
        const bigIntString = await this.runCommand(`prime -bits ${bitLength} -generate -safe`);
        return BigInt(bigIntString);
    }
    async runCommand(openSslCommand) {
        if (!this.isInitialized) {
            throw new Error('The OpenSSl class is not initialized! Please call OpenSSL.init().');
        }
        const command = Array.isArray(openSslCommand) ? openSslCommand : openSslCommand.split(/[\s]{1,}/g).filter(Boolean);
        const wasi = new wasi_1.WASI({
            args: command,
        });
        // Instantiate the WASI module
        // cannot use wasi.instantiate(module, {}); due to the size of the module
        const instance = await WebAssembly.instantiate(this.waModule, {
            ...wasi.getImports(this.waModule),
        });
        // Run the start function
        wasi.start(instance);
        return wasi.getStdoutString();
    }
}
exports.OpenSSL = OpenSSL;
