// @bitgo/sdk-coin-sui/dist/src/suiToken.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SuiToken = void 0;
const statics_1 = require("@bitgo/statics");
const sui_1 = require("./sui");
class SuiToken extends sui_1.Sui {
    constructor(bitgo, tokenConfig) {
        const staticsCoin = tokenConfig.network === 'Mainnet' ? statics_1.coins.get('sui') : statics_1.coins.get('tsui');
        super(bitgo, staticsCoin);
        this.tokenConfig = tokenConfig;
    }
    static createTokenConstructor(config) {
        return (bitgo) => new SuiToken(bitgo, config);
    }
    static createTokenConstructors() {
        const tokensCtors = [];
        for (const token of [...statics_1.tokens.bitcoin.sui.tokens, ...statics_1.tokens.testnet.sui.tokens]) {
            const tokenConstructor = SuiToken.createTokenConstructor(token);
            tokensCtors.push({ name: token.type, coinConstructor: tokenConstructor });
        }
        return tokensCtors;
    }
    get name() {
        return this.tokenConfig.name;
    }
    get coin() {
        return this.tokenConfig.coin;
    }
    get network() {
        return this.tokenConfig.network;
    }
    get packageId() {
        return this.tokenConfig.packageId;
    }
    get module() {
        return this.tokenConfig.module;
    }
    get symbol() {
        return this.tokenConfig.symbol;
    }
    get decimalPlaces() {
        return this.tokenConfig.decimalPlaces;
    }
    getChain() {
        return this.tokenConfig.type;
    }
    getBaseChain() {
        return this.coin;
    }
    getFullName() {
        return 'Sui Token';
    }
    getBaseFactor() {
        return Math.pow(10, this.tokenConfig.decimalPlaces);
    }
}
exports.SuiToken = SuiToken;
