// @bitgo/sdk-core/dist/src/bitgo/lightning/lightningUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.unwrapLightningCoinSpecific = exports.getUtxolibNetworkName = exports.isValidLightningNetwork = exports.isValidLightningNetworkName = exports.getLightningCoinName = exports.getLightningNetwork = exports.lightningNetworkName = void 0;
const statics = __importStar(require("@bitgo/statics"));
const utxolib = __importStar(require("@bitgo/utxo-lib"));
exports.lightningNetworkName = ['bitcoin', 'testnet'];
/**
 * Get the utxolib network for a lightning network.
 */
function getLightningNetwork(networkName) {
    return utxolib.networks[networkName];
}
exports.getLightningNetwork = getLightningNetwork;
/**
 * Get the lightning coin name for a utxolib network.
 */
function getLightningCoinName(network) {
    return network === utxolib.networks.bitcoin ? 'lnbtc' : 'tlnbtc';
}
exports.getLightningCoinName = getLightningCoinName;
/**
 * Checks if the network name is a valid lightning network name.
 */
function isValidLightningNetworkName(networkName) {
    return exports.lightningNetworkName.includes(networkName);
}
exports.isValidLightningNetworkName = isValidLightningNetworkName;
/**
 * Checks if the network is a valid lightning network.
 */
function isValidLightningNetwork(network) {
    return utxolib.isValidNetwork(network) && isValidLightningNetworkName(utxolib.getNetworkName(network));
}
exports.isValidLightningNetwork = isValidLightningNetwork;
/**
 * Returns the utxolib network name for a lightning coin.
 */
function getUtxolibNetworkName(coinName) {
    const coin = statics.coins.get(coinName);
    return coin instanceof statics.LightningCoin ? coin.network.utxolibName : undefined;
}
exports.getUtxolibNetworkName = getUtxolibNetworkName;
/**
 * Returns coin specific data for a lightning coin.
 */
function unwrapLightningCoinSpecific(obj, coinSpecificPath) {
    if (coinSpecificPath !== 'lnbtc' && coinSpecificPath !== 'tlnbtc') {
        throw new Error(`invalid coinSpecificPath ${coinSpecificPath} for lightning coin`);
    }
    if (coinSpecificPath === 'lnbtc' && 'lnbtc' in obj) {
        return obj.lnbtc;
    }
    if (coinSpecificPath === 'tlnbtc' && 'tlnbtc' in obj) {
        return obj.tlnbtc;
    }
    throw new Error('invalid lightning coin specific');
}
exports.unwrapLightningCoinSpecific = unwrapLightningCoinSpecific;
