// @bitgo/sdk-lib-mpc/dist/src/tss/ecdsa-dkls/dsg.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Dsg = void 0;
const types_1 = require("./types");
const cbor_x_1 = require("cbor-x");
class Dsg {
    constructor(keyShare, partyIdx, derivationPath, messageHash, dklsWasm) {
        this.dsgState = types_1.DsgState.Uninitialized;
        this.partyIdx = partyIdx;
        this.keyShareBytes = keyShare;
        this.derivationPath = derivationPath;
        this.messageHash = messageHash;
        this.dklsWasm = dklsWasm !== null && dklsWasm !== void 0 ? dklsWasm : null;
    }
    _restoreSession() {
        if (!this.dsgSession) {
            this.dsgSession = this.getDklsWasm().SignSession.fromBytes(this.dsgSessionBytes);
        }
    }
    _deserializeState() {
        if (!this.dsgSession) {
            throw Error('Session not intialized');
        }
        const round = (0, cbor_x_1.decode)(this.dsgSession.toBytes()).round;
        switch (round) {
            case 'WaitMsg1':
                this.dsgState = types_1.DsgState.Round1;
                break;
            case 'WaitMsg2':
                this.dsgState = types_1.DsgState.Round2;
                break;
            case 'WaitMsg3':
                this.dsgState = types_1.DsgState.Round3;
                break;
            case 'Ended':
                this.dsgState = types_1.DsgState.Complete;
                break;
            default:
                this.dsgState = types_1.DsgState.InvalidState;
                throw Error(`Invalid State: ${round}`);
        }
    }
    async loadDklsWasm() {
        if (!this.dklsWasm) {
            this.dklsWasm = await Promise.resolve().then(() => __importStar(require('@silencelaboratories/dkls-wasm-ll-node')));
        }
    }
    getDklsWasm() {
        if (!this.dklsWasm) {
            throw Error('DKLS wasm not loaded');
        }
        return this.dklsWasm;
    }
    /**
     * Returns the current DSG session as a base64 string.
     * @returns {string} - base64 string of the current DSG session
     */
    getSession() {
        return Buffer.from(this.dsgSessionBytes).toString('base64');
    }
    /**
     * Sets the DSG session from a base64 string.
     * @param {string} session - base64 string of the DSG session
     */
    async setSession(session) {
        this.dsgSession = undefined;
        if (!this.dklsWasm) {
            await this.loadDklsWasm();
        }
        const sessionBytes = new Uint8Array(Buffer.from(session, 'base64'));
        const round = (0, cbor_x_1.decode)(sessionBytes).round;
        switch (true) {
            case round === 'WaitMsg1':
                this.dsgState = types_1.DsgState.Round1;
                break;
            case round === 'WaitMsg2':
                this.dsgState = types_1.DsgState.Round2;
                break;
            case round === 'WaitMsg3':
                this.dsgState = types_1.DsgState.Round3;
                break;
            case 'WaitMsg4' in round:
                this.dsgState = types_1.DsgState.Round4;
                break;
            default:
                throw Error(`Invalid State: ${round}`);
        }
        this.dsgSessionBytes = sessionBytes;
    }
    async init() {
        var _a;
        if (this.dsgState !== types_1.DsgState.Uninitialized) {
            throw Error('DSG session already initialized');
        }
        if (!this.dklsWasm) {
            await this.loadDklsWasm();
        }
        if (typeof window !== 'undefined' &&
            /* checks for electron processes */
            !window.process &&
            !((_a = window.process) === null || _a === void 0 ? void 0 : _a['type'])) {
            /* This is only needed for browsers/web because it uses fetch to resolve the wasm asset for the web */
            const initDkls = await Promise.resolve().then(() => __importStar(require('@silencelaboratories/dkls-wasm-ll-web')));
            await initDkls.default();
        }
        const { Keyshare, SignSession } = this.getDklsWasm();
        const keyShare = Keyshare.fromBytes(this.keyShareBytes);
        if (keyShare.partyId !== this.partyIdx) {
            throw Error(`Party index: ${this.partyIdx} does not match key share partyId: ${keyShare.partyId} `);
        }
        this.dsgSession = new SignSession(keyShare, this.derivationPath);
        try {
            const payload = this.dsgSession.createFirstMessage().payload;
            this._deserializeState();
            this.dsgSessionBytes = this.dsgSession.toBytes();
            this.dsgSession = undefined;
            return {
                payload: payload,
                from: this.partyIdx,
            };
        }
        catch (e) {
            throw Error(`Error while creating the first message from party ${this.partyIdx}: ${e}`);
        }
    }
    get signature() {
        if (!this._signature) {
            throw Error('Can not request signature. Signature not produced yet.');
        }
        return this._signature;
    }
    /**
     * Ends the DSG session by freeing any heap allocations from wasm. Note that the session is freed if a signature is produced.
     */
    endSession() {
        if (this._signature) {
            new Error('Session already ended because combined signature was produced.');
        }
        if (this.dsgSession) {
            this.dsgSession.free();
        }
        this.dsgState = types_1.DsgState.Uninitialized;
    }
    /**
     * Proccesses incoming messages to this party in the DKLs DSG protocol and
     * produces messages from this party to other parties for the next round.
     * @param messagesForIthRound - messages to process the current round
     * @returns {DeserializedMessages} - messages to send to other parties for the next round
     */
    handleIncomingMessages(messagesForIthRound) {
        let nextRoundMessages = [];
        let nextRoundDeserializedMessages = { broadcastMessages: [], p2pMessages: [] };
        this._restoreSession();
        if (!this.dsgSession) {
            throw Error('Session not initialized');
        }
        const { Message } = this.getDklsWasm();
        try {
            if (this.dsgState === types_1.DsgState.Round4) {
                this.dsgState = types_1.DsgState.Complete;
                const combineResult = this.dsgSession.combine(messagesForIthRound.broadcastMessages.map((m) => new Message(m.payload, m.from, undefined)));
                this._signature = {
                    R: combineResult[0],
                    S: combineResult[1],
                };
                return { broadcastMessages: [], p2pMessages: [] };
            }
            else {
                nextRoundMessages = this.dsgSession.handleMessages(messagesForIthRound.broadcastMessages
                    .map((m) => new Message(m.payload, m.from, undefined))
                    .concat(messagesForIthRound.p2pMessages.map((m) => new Message(m.payload, m.from, m.to))));
            }
            if (this.dsgState === types_1.DsgState.Round3) {
                nextRoundMessages = [this.dsgSession.lastMessage(this.messageHash)];
                this.dsgState = types_1.DsgState.Round4;
                return {
                    broadcastMessages: [
                        {
                            payload: nextRoundMessages[0].payload,
                            from: nextRoundMessages[0].from_id,
                            signatureR: (0, cbor_x_1.decode)(this.dsgSession.toBytes()).round.WaitMsg4.r,
                        },
                    ],
                    p2pMessages: [],
                };
            }
            else {
                // Update round data.
                this._deserializeState();
            }
            nextRoundDeserializedMessages = {
                p2pMessages: nextRoundMessages
                    .filter((m) => m.to_id !== undefined)
                    .map((m) => {
                    if (m.to_id === undefined) {
                        throw Error('Invalid P2P message, missing to_id.');
                    }
                    const p2pReturn = {
                        payload: m.payload,
                        from: m.from_id,
                        to: m.to_id,
                    };
                    return p2pReturn;
                }),
                broadcastMessages: nextRoundMessages
                    .filter((m) => m.to_id === undefined)
                    .map((m) => {
                    const broadcastReturn = {
                        payload: m.payload,
                        from: m.from_id,
                    };
                    return broadcastReturn;
                }),
            };
        }
        catch (e) {
            if (e.message.startsWith('Abort the protocol and ban')) {
                throw Error('Signing aborted. Please stop all transaction signing from this wallet and contact support@bitgo.com.');
            }
            throw Error(`Error while creating messages from party ${this.partyIdx}, round ${this.dsgState}: ${e}`);
        }
        finally {
            nextRoundMessages.forEach((m) => m.free());
            // Session is freed when combine is called.
            if (this.dsgState !== types_1.DsgState.Complete) {
                this.dsgSessionBytes = this.dsgSession.toBytes();
                this.dsgSession = undefined;
            }
        }
        return nextRoundDeserializedMessages;
    }
}
exports.Dsg = Dsg;
