// @bitgo/sdk-coin-ton/dist/src/lib/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Utils = void 0;
const sdk_core_1 = require("@bitgo/sdk-core");
const tonweb_1 = __importDefault(require("tonweb"));
class Utils {
    /** @inheritdoc */
    isValidAddress(address) {
        try {
            if (address.length != 48) {
                return false;
            }
            Buffer.from(address, 'base64');
            return true;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidBlockId(hash) {
        try {
            return Buffer.from(hash, 'base64').length === 32;
        }
        catch (e) {
            return false;
        }
    }
    /** @inheritdoc */
    isValidPrivateKey(key) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    isValidPublicKey(key) {
        return (0, sdk_core_1.isValidEd25519PublicKey)(key);
    }
    /** @inheritdoc */
    isValidSignature(signature) {
        throw new Error('Method not implemented.');
    }
    /** @inheritdoc */
    isValidTransactionId(txId) {
        try {
            return Buffer.from(txId, 'base64').length === 32;
        }
        catch (e) {
            return false;
        }
    }
    async getAddressFromPublicKey(publicKey, bounceable = true) {
        const tonweb = new tonweb_1.default(new tonweb_1.default.HttpProvider(''));
        const WalletClass = tonweb.wallet.all['v4R2'];
        const wallet = new WalletClass(tonweb.provider, {
            publicKey: tonweb_1.default.utils.hexToBytes(publicKey),
            wc: 0,
        });
        const address = await wallet.getAddress();
        return address.toString(true, true, bounceable);
    }
    getAddress(address, bounceable = true) {
        if (bounceable) {
            return new tonweb_1.default.Address(address).isBounceable
                ? address
                : new tonweb_1.default.Address(address).toString(true, true, bounceable);
        }
        else {
            return new tonweb_1.default.Address(address).isBounceable
                ? new tonweb_1.default.Address(address).toString(true, true, bounceable)
                : address;
        }
    }
    async getMessageHashFromData(data) {
        const cell = tonweb_1.default.boc.Cell.oneFromBoc(tonweb_1.default.utils.base64ToBytes(data));
        // this is need to be confirmed by ton team
        const message = cell.refs[0].refs[0];
        const hash = tonweb_1.default.utils.bytesToBase64(await message.hash());
        return hash.toString();
    }
}
exports.Utils = Utils;
const utils = new Utils();
exports.default = utils;
